---
title: "TRAFTAC"
date: "11/2/2023"
output: html_document
---

```{r setup}
suppressPackageStartupMessages({
  library(BiocParallel)
  library(chromVAR)
  library(SummarizedExperiment)
  library(motifmatchr)
  library(monaLisa)
  library(sechm)
  library(limma)
  library(ggplot2)
  library(edgeR)
})
source("../Scripts/getpmoi.R")
source("../Scripts/runStabSel.R")
source("../Scripts/GCnorm.R")
source("../Scripts/differentialAccessibility.R")
genome <- Rsamtools::FaFile("/reference/Homo_sapiens/Ensembl/GRCh38.p10/Sequence/WholeGenomeFasta/Homo_sapiens.GRCh38.dna.toplevel.fa")
```

## chromVAR

```{r, eval=FALSE}
se <- readRDS("peakCounts.SE.rds")

mm <- model.matrix(~se$condition)
se <- se[edgeR::filterByExpr(assay(se),mm),]
row.names(se) <- as.character(granges(se))

dds <- calcNormFactors(DGEList(assay(se), group=se$condition))
dds <- estimateDisp(dds,mm)
res <- as.data.frame(topTags(glmLRT(glmFit(dds,mm),coef="se$conditionNFkB"), Inf))
rowData(se)$DEA.nfkp <- res[row.names(se),]
saveRDS(se, "peakCounts.SE.rds")

motifs <- getNonRedundantMotifs(species="Hsapiens")
moi <- motifmatchr::matchMotifs(motifs, subject=se, genome=genome)
saveRDS(moi, "motifMatches.rds")
set.seed(1234)
dev <- chromVAR::computeDeviations(object=se, annotations=moi,
                                   background_peaks=getBackgroundPeaks(se, niterations=2000))
q <- preprocessCore::normalize.quantiles(assay(dev))
dimnames(q) <- dimnames(dev)
assays(dev)$qt <- q
dev2 <- dev[!grepl(":",row.names(dev),fixed=TRUE) & !grepl("var.2",row.names(dev),fixed=TRUE),]
fit <- eBayes(lmFit(assays(dev2)$z, mm))
res <- as.data.frame(topTable(fit, coef="se$conditionNFkB", number = Inf))
res$TF <- row.names(res)
rowData(dev) <- res[row.names(dev),]
saveRDS(dev, file="deviations.SE.rds")
```

```{r}
dev <- readRDS("deviations.SE.rds")
cv <- as.data.frame(rowData(dev))
cv$absMaxDeviation <- matrixStats::rowMaxs(abs(assays(dev)$deviations))
cv$absMaxZ <- matrixStats::rowMaxs(abs(assays(dev)$z))
cv <- cv[order(cv$P.Value),]
cv2 <- cv[which(abs(cv$logFC)>4.8 & cv$adj.P.Val<0.2 & cv$absMaxDeviation>0.0122),]
ggplot(cv, aes(absMaxDeviation, logFC, colour=-log10(adj.P.Val), label=TF)) +
  geom_point() + scale_color_viridis_c(option="A",direction=-1) + theme_bw() +
  geom_point(data=cv[grep("NFKB",row.names(cv)),], size=5, shape=1, alpha=1, colour="darkred") +
  ggrepel::geom_text_repel(data=cv2, min.segment.length=0, max.overlaps=30, show.legend=FALSE) +
  labs(x="Maximum absolute chromVAR deviation", y="Motif log2(foldchange)")
```

```{r}
metadata(dev)$anno_colors <- list(condition=c("Ctrl"="lightgrey", "NFkB"="darkred"))
sechm(dev, row.names(cv2), assayName = "z", do.scale = TRUE)
```


## monaLisa

```{r, eval=FALSE}
se <- readRDS("peakCounts.SE.rds")
se <- keepStandardChromosomes(se, "Homo_sapiens", pruning.mode="coarse")
da <- rowData(se)$DEA.nfkp
motifs <- getNonRedundantMotifs(species="Hsapiens", format="PWMatrix")
peaks <- granges(se)
names(peaks) <- row.names(se)

peaks$logFC <- jitter(da[names(peaks),"logFC"], 0.001) # To avoid non-unique breaks
fc2 <- peaks$logFC[which(abs(peaks$logFC)>=0.2)]
bins <- monaLisa::bin(x=peaks$logFC, 
                      binmode = "equalN", 
                      nElements = ceiling(length(fc2)/10),
                      minAbsX = 0.2)
  
#plotBinDensity(peaks$logFC, bins)

DARseqs <- getSeq(genome, peaks)
names(DARseqs) <- names(peaks)
  
plotBinDiagnostics(seqs=DARseqs, bins=bins, aspect="dinucfreq")
# strong GC-bias...
  
mlse <- calcBinnedMotifEnrR(seqs = DARseqs, 
                            bins = bins, 
                            pwmL = motifs, 
                            BPPARAM = BiocParallel::MulticoreParam(8),
                            background = "otherBins")
  
simes <- function(pval){ 
  min((length(pval)*pval[order(pval)])/seq(from=1, to=length(pval), by=1))
}
MLp <- 10^-assays(mlse)$negLog10P[,-which(colData(mlse)$bin.nochange)]
rowData(mlse)$simes.p <- apply(MLp, 1, simes)
rowData(mlse)$simes.padj <- p.adjust(rowData(mlse)$simes.p, method="fdr")
mlse <- mlse[order(rowData(mlse)$simes.p),]
saveRDS(mlse, "monaLisa.SE.rds")

mlse <- calcBinnedMotifEnrR(seqs = DARseqs, 
                            bins = bins, 
                            pwmL = motifs, 
                            BPPARAM = BiocParallel::MulticoreParam(8),
                            background = "zeroBin")

MLp <- 10^-assays(mlse)$negLog10P[,-which(colData(mlse)$bin.nochange)]
rowData(mlse)$simes.p <- apply(MLp, 1, simes)
rowData(mlse)$simes.padj <- p.adjust(rowData(mlse)$simes.p, method="fdr")
mlse <- mlse[order(rowData(mlse)$simes.p),]
saveRDS(mlse, "monaLisa.vsZero.SE.rds")
```

```{r, eval=FALSE}
se <- readRDS("peakCounts.SE.rds")
se <- keepStandardChromosomes(se, "Homo_sapiens", pruning.mode="coarse")
da <- dATestedgeR(se[,1:2], se[,3:4], norm.method="GCSmoothQuantile")
rowData(se)$DEA.withGCsmoothQ <- da

motifs <- getNonRedundantMotifs(species="Hsapiens", format="PWMatrix")
peaks <- granges(se)
names(peaks) <- row.names(se)

peaks$logFC <- jitter(da[names(peaks),"logFC"], 0.001) # To avoid non-unique breaks
fc2 <- peaks$logFC[which(abs(peaks$logFC)>=0.2)]
bins <- monaLisa::bin(x=peaks$logFC, 
                      binmode = "equalN", 
                      nElements = ceiling(length(fc2)/10),
                      minAbsX = 0.2)
  
#plotBinDensity(peaks$logFC, bins)

DARseqs <- getSeq(genome, peaks)
names(DARseqs) <- names(peaks)
  
plotBinDiagnostics(seqs=DARseqs, bins=bins, aspect="dinucfreq")
# strong GC-bias...
  
mlse <- calcBinnedMotifEnrR(seqs = DARseqs, 
                            bins = bins, 
                            pwmL = motifs, 
                            BPPARAM = BiocParallel::MulticoreParam(8),
                            background = "otherBins")
  

MLp <- 10^-assays(mlse)$negLog10P[,-which(colData(mlse)$bin.nochange)]
rowData(mlse)$simes.p <- apply(MLp, 1, simes)
rowData(mlse)$simes.padj <- p.adjust(rowData(mlse)$simes.p, method="fdr")
mlse <- mlse[order(rowData(mlse)$simes.p),]
saveRDS(mlse, "monaLisa.GCsmoothQ.SE.rds")

mlse <- calcBinnedMotifEnrR(seqs = DARseqs, 
                            bins = bins, 
                            pwmL = motifs, 
                            BPPARAM = BiocParallel::MulticoreParam(8),
                            background = "zeroBin")

MLp <- 10^-assays(mlse)$negLog10P[,-which(colData(mlse)$bin.nochange)]
rowData(mlse)$simes.p <- apply(MLp, 1, simes)
rowData(mlse)$simes.padj <- p.adjust(rowData(mlse)$simes.p, method="fdr")
mlse <- mlse[order(rowData(mlse)$simes.p),]
saveRDS(mlse, "monaLisa.GCsmoothQ.vsZero.SE.rds")
```


```{r}
mlse <- readRDS("monaLisa.GCsmoothQ.SE.rds")

plotMotifHeatmaps(x=head(mlse,10), which.plots=c("log2enr", "negLog10P"), 
                  width = 2.0, cluster = TRUE)
```

## Comparison of the two

```{r}
cv$ml.p <- rowData(mlse)[row.names(cv),"simes.p"]
ggplot(cv, aes(-log10(P.Value), -log10(ml.p), colour=abs(logFC))) + geom_point() + 
  scale_colour_viridis_c(direction=-1) + theme_bw() + 
  labs(x="chromVAR -log10(p-value)", y="monaLisa -log10(p-value)")
```


## Stability selection

```{r}
se <- readRDS("peakCounts.SE.rds")
se <- keepStandardChromosomes(se, "Homo_sapiens", pruning.mode="coarse")
da <- rowData(se)$DEA.nfkp
motifs <- getNonRedundantMotifs(species="Hsapiens", format="PWMatrix")
peaks <- granges(se)
names(peaks) <- row.names(se)
peaks$logFC <- jitter(da[names(peaks),"logFC"], 0.001) # To avoid non-unique breaks
library(BiocParallel)
ss <- runStabSel(da, motifs, peaks, genome)
ss <- ss$res
saveRDS(ss, file="stabSel.rds")
selected <- intersect(colnames(ss)[which(ss$selProb>=0.9)],row.names(mlse))
saveRDS(selected, file="stabSel.selected.rds")

plotMotifHeatmaps(x = mlse[selected,],
                  which.plots = c("log2enr", "negLog10P"), 
                  width = 2.0, cluster = TRUE, 
                  show_motif_GC = TRUE, show_seqlogo = TRUE)
```

```{r, fig.width=8, fig.height=5}
uMo <- setdiff(intersect(union(selected, row.names(cv2)), row.names(mlse)), "NF2L1") # infinite value for that genes causes error
hl <- plotMotifHeatmaps(x = mlse[uMo,],
                  which.plots = c("log2enr"), width.seqlogo=2,
                  width = 1.25, cluster = TRUE, doPlot=FALSE, column_title_gp=gpar(fontsize=11),
                  show_motif_GC = TRUE, show_seqlogo = TRUE,
                  row_names_gp=gpar(fontsize=9))
draw(
  Reduce("+",hl) + sechm(dev, uMo, assayName = "z", do.scale=TRUE, 
                       column_title_gp=gpar(fontsize=11), column_title="chromVAR",
                       width=unit(0.7, "in"), row_names_gp=gpar(fontsize=9)),
  merge=TRUE)
```

## smoothGC+mlm

```{r}
se <- readRDS("peakCounts.SE.rds")
moi <- readRDS("motifMatches.rds")
assays(se)$gcnorm <- GCSmoothQuantile(se, bio="condition")
```

```{r}
#' mlmFast
#'
#' @param accmat A matrix of peak (rows) accessibility per sample/cell (columns)
#' @param annotation A matrix of motif (columns) matches per peak (rows)
#' @param useIntercept Logical; whether to use an intercept in the mode
#' @param poisson Logical; whether to use poisson regression (assumes `accmat`
#'   is a count matrix).
#' @param BPPARAM BiocParallel param for multithreading.
#'
#' @return A matrix (of `ncol(annotation)` rows and `ncol(accmat)` columns) with
#'   the activity scores (model t-values) of each motif in each sample.
#'   
#' @details 
#' Regresses each column of `accmat` on `annotation`, and uses the coefficients'
#' t-values as activity scores.
#' 
#' @export
mlmFast <- function(accmat, annotation, useIntercept=TRUE, poisson=FALSE,
                    BPPARAM=BiocParallel::SerialParam()){
  stopifnot(is.matrix(accmat) && is.matrix(annotation))
  if(useIntercept) annotation <- cbind(rep(1L,nrow(annotation)), annotation)
  res <- bplapply(seq_len(ncol(accmat)), BPPARAM=BPPARAM, FUN=function(i){
    if(!isTRUE(poisson)){
      mod <- RcppArmadillo::fastLmPure(annotation, accmat[,i])
      tvals <- mod$coefficients/mod$stderr
    }else{
      if(FALSE && require("Rfast", quietly=TRUE, include.only="glm_poisson")){
        mod <- glm_poisson(a, y[,1], full=TRUE)$info
      }else{
        mod <- glm(accmat[,i]~0+annotation, family="poisson")
        mod <- coef(summary(mod))
      }
      tvals <- mod[,1]/mod[,2]
    }
    tvals
  })
  res <- matrix(unlist(res), nrow=ncol(annotation))
  row.names(res) <- colnames(annotation)
  if(useIntercept) res <- res[-1,,drop=FALSE]
  colnames(res) <- colnames(accmat)
  res
}
```

```{r}
e <- mlmFast(assays(se)$gcnorm, as.matrix(assay(moi)), BPPARAM=MulticoreParam(4))
mse <- SummarizedExperiment(list(lm=e), colData=colData(se))
mm <- model.matrix(~se$condition)
fit <- eBayes(lmFit(assay(mse), mm))
res <- as.data.frame(topTable(fit, number=Inf))
res$TF <- row.names(res)
head(res)
rowData(mse) <- res[row.names(mse),]
rowData(mse)$meanAbsT <- rowMeans(abs(assay(mse)))
saveRDS(mse,"gcnorm.mlm.SE.rds")
sechm(mse, head(row.names(res),20), do.scale = TRUE, left_annotation="meanAbsT")
```

## insertion model

```{r}
bams <- list.files("/mnt/germain/datasets/2023_nfkbTraftac_ATAC/aligned", pattern="bam$", full=TRUE)
motifs <- getNonRedundantMotifs(species="Hsapiens")
moi <- motifmatchr::matchMotifs(motifs, subject=readRDS("peakCounts.SE.rds"), genome=genome, out="positions")
moi <- unlist(GRangesList(lapply(names(moi), FUN=function(x){ y <- moi[[x]]; y$motif_id <- x; y })))
moi$motif_id <- factor(moi$motif_id)
moi <- keepSeqlevels(moi,seqlevelsInUse(moi),pruning.mode = "coarse")
source("../Scripts/runBaGFoot.R")
se.mba <- getModelBasedActivity(bams, moi, paired=TRUE, BPPARAM=MulticoreParam(4, progress=TRUE))
dds <- calcNormFactors(DGEList(assay(se.mba)))
groups <- rep(LETTERS[1:2],each=2)
design <- model.matrix(~groups)
fit <- eBayes(lmFit(voom(dds, design), design))
topTFs <- as.data.frame(topTable(fit, number = Inf))
rowData(se.mba)$DA <- topTFs[row.names(se.mba),]
saveRDS(se.mba, "insertionModel.SE.rds")
```


## Checking the insertion profile of some candidates

```{r, fig.width=7, fig.height=4}
motifs <- getNonRedundantMotifs(species="Hsapiens")[c("NFKB1")]
moi <- motifmatchr::matchMotifs(motifs, subject=se, genome=genome, out="positions")
moi <- moi[[1]]
names(moi) <- as.character(granges(moi))
moi <- keepSeqlevels(moi,seqlevelsInUse(moi),pruning.mode = "coarse")
tracks <- list.files("/mnt/plger/datasets/2023_nfkbTraftac_ATAC/tracks2/", pattern="\\.bw", full=TRUE)
sm <- signal2Matrix(tracks, regions=moi, extend=250L, w=1L)
nf <- epiwraps::getNormFactors(tracks)
sm <- epiwraps::renormalizeSignalMatrices(sm, scaleFactors=nf)
d1 <- epiwraps::meltSignals(sm, trim = 0.99999)

tracks <- list.files("/mnt/plger/datasets/2023_nfkbTraftac_ATAC/biasCorrected/", pattern="_corrected\\.bw", full=TRUE)
sm <- signal2Matrix(tracks, regions=moi, extend=250L, w=1L)
nf <- epiwraps::getNormFactors(tracks)
sm <- epiwraps::renormalizeSignalMatrices(sm, scaleFactors=nf)
d2 <- epiwraps::meltSignals(sm, trim = 0.99999)

d1$type <- "uncorrected"
d2$type <- "bias-corrected"
levels(d2$sample) <- gsub("_corrected","",levels(d2$sample))

d <- rbind(d1,d2)
d$condition <- d$sample
levels(d$condition) <- ifelse(grepl("Ctrl",levels(d$sample)), "CTRL", "NFKB-KD")
d$type <- factor(d$type, unique(d$type))
levels(d$type)[2] <- "bias-corrected (TOBIAS)"
saveRDS(d, "Nfkb_profile.rds")
ggplot(d, aes(position, mean, group=sample, colour=condition)) + geom_line(linewidth=0.75) + 
  facet_wrap(~type, scales="free_y", nrow=2) + theme_bw() + xlim(-100,100) + 
  labs(x="Position around motif center",
       y="Normalized insertion sites")
```


```{r}
mo <- t(as.matrix(motifs[[1]])+1L)
mo <- t(mo/rowSums(mo))
genome <- BSgenome.Hsapiens.UCSC.hg38
seqlevelsStyle(genome) <- "ensembl"
ATACseqQC::factorFootprints("/mnt/plger/datasets/2023_nfkbTraftac_ATAC/aligned/Ctrl_PG_1.bam", pfm=mo, genome=genome)
```



```{r}
e <- sapply(setNames(names(ds),names(ds)), FUN=function(d){
  ml <- readRDS(paste0("../fullFrags/",d,"/runATAC_results/raw/ML_raw.rds"))[[1]]
  ml <- ml[order(getMLglobalP(ml)),]
  rank1 <- head(which(row.names(ml) %in% ds[[d]]$truth),1)
  #ml <- ml[order(getMLglobalP(ml, method="weighted.cauchy")),]
  ml <- ml[order(getMLglobalP(ml, method="wgeomean")),]
  rank2 <- head(which(row.names(ml) %in% ds[[d]]$truth),1)
  c(simes=rank1, wcauchy=rank2)
})
```

