---
title: "On the identification of differentially-active transcription fators from ATAC-seq data"
subtitle: "Supplementary Figures"
date: "`r format(Sys.time(), '%d %B, %Y')`"
author: "Gerbaldo, Sonder et al."
output:
  pdf_document:
    fig_width: 8
    fig_height: 6
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, echo=FALSE)
FIG_NB <- 0
FIG_STRING <- "Supplementary Figure "
getFigNb <- function(increment=FALSE){
  if(increment) FIG_NB <<- FIG_NB + 1
  paste0(FIG_STRING,FIG_NB)
}
```

```{r packages}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(ggplot2)
  library(cowplot)
})
source("../Scripts/plot.R")
source("../Scripts/runMethods.R")
```



# `r getFigNb(TRUE)`

```{r DAvolcanos}
das2 <- readRDS("../misc/allDAres.rds")
nsig <- dplyr::bind_rows(lapply(split(das2, das2$Dataset), FUN=function(x){
  data.frame(n=nrow(x), sig=sum(x$FDR<0.05,na.rm=TRUE),
             unadj=sum(x$PValue<0.05,na.rm=TRUE),
             unadjProp=sum(x$PValue<0.05,na.rm=TRUE)/nrow(x))
}), .id="Dataset")
nsig$label <- paste0("q<0.05: ", nsig$sig, "\np<0.05: ", round(100*nsig$unadjProp), "%")
p1 <- ggplot(das2, aes(logFC, -log10(FDR))) + facet_wrap(~Dataset) + 
  geom_hline(yintercept = -log10(0.05), linetype="dashed", colour="grey") + 
  ggrastr::geom_point_rast(size=0.5) + theme_bw() + scale_y_sqrt() +
  geom_text(data=nsig, x=-4, y=max(sqrt(-log10(das2$FDR))), vjust=1, aes(label=label), size=3)
```

```{r TFmotifInfo, fig.width=9, fig.height=3}
peakr <- readRDS("../misc/motifPeakData.rds")
p2 <- cowplot::plot_grid(
  ggplot(peakr, aes(Dataset, atacPeakLog2EnrichmentWithinMotifs, fill=Motif, label=Motif, Group=Motif)) + 
  geom_col(position="dodge") + geom_text(position = position_dodge(width = .9), angle=90, hjust=0, size=3) + 
  ylab("log2(enrichment) of ChIP peaks\nin motif-containing ATAC peaks") + ylim(c(-0.3,5)) +
  theme_bw() + theme(legend.position = "none", axis.text.x=element_text(angle=45, hjust=1, vjust=1)),
  ggplot(peakr, aes(Dataset, ChIPpeaksWithMotif, fill=Motif, label=Motif, Group=Motif)) + 
  geom_col(position="dodge") + geom_text(position = position_dodge(width = .9), angle=90, hjust=0, size=3) + 
  ylab("Proportion of ChIP peaks\ncontaining the motif") + ylim(0,0.5) +
  theme_bw() + theme(legend.position = "none", axis.text.x=element_text(angle=45, hjust=1, vjust=1)),
  nrow=1, labels=LETTERS[2:3], scale=0.95)
#p2
```

```{r SF1, fig.height=10, fig.width=9}
cowplot::plot_grid(
  p1, p2, nrow=2, rel_heights=c(2,1), labels=c("A",NA), scale=c(0.95,1)
)
```

**`r getFigNb()` : Overview of the benchmark datasets and TFs. **
**A:** Volcano plot of the peak-level differential accessibility analyses, 
illustrating the extent and significance of changes upon treatment in each dataset.
The dashed line represents a 0.05 FDR threshold, and indicated are the number of
ATAC peaks passing this threshold, as well as the proportion of peaks with an 
uncorrected p-value lower than 0.05. 
**B:** Enrichment for experimental binding sites of the factor (main TF after 
which the datasets are named) in ATAC peaks containing the respective motifs, 
versus all ATAC peaks. 
**C:** Proportion of ChIPseq peaks (overlapping ATAC peaks) that contain the 
respective motif.


\newpage





# `r getFigNb(TRUE)`

```{r}
cors <- readRDS("../misc/CV_repro_cors.rds")
means <- aggregate(cors$correlation, cors[,1:2], FUN=mean)
se <- aggregate(cors$correlation, cors[,1:2], FUN=sd)
se$x <- se$x/sqrt(7)
means$upper <- means$x+se$x
means$lower <- means$x-se$x
p1 <- ggplot(means, aes(niter, x, group=dataset)) + 
  geom_ribbon(aes(ymax=upper, ymin=lower, fill=dataset), alpha=0.15) + 
  geom_line(aes(colour=dataset), linewidth=1) + theme_bw() +
  labs(y="Correlation, across random seeds, of\nthe ranks of the union of top 20 motifs",
       x="Number of chromVAR background iterations\n(default 50)",
       title="ChromVAR ranking reproducibility") +
  theme(legend.position="none", title=element_text(size=10)) + 
  coord_cartesian(expand=FALSE, ylim=c(0.5,1))
```


```{r}
se <- readRDS("../fullFrags/BANP/runATAC_results/raw/MBA_raw.rds")$obj1
gl <- metadata(se)$globalProfiles
row.names(gl) <- row.names(se)
colnames(gl) <- seq(from=-200,to=200)[-201]
m <- reshape2::melt(gl, varnames=c("Motif","Position"))
m$Position <- as.integer(m$Position)

p2 <- ggplot(m[m$Motif %in% c("FOXO1","SOX2", "CTCF"),], aes(Position, value, colour=Motif)) +
  geom_line(linewidth=0.9) + theme_bw() + 
  labs(x="Position around motif center", y="Weight",
       title="Position weights based on insertion sites\naround the motif in the whole dataset") +
  theme(legend.position=c(0.84,0.78), panel.grid.major=element_blank(), title=element_text(size=10),
        panel.grid.minor=element_blank()) + coord_cartesian(expand=FALSE)
```


```{r, fig.width=8, fig.height=5}
ints <- readRDS("../Scripts/allInteractors.rds")
resl <- lapply(c("chromVAR-adjusted"="CVqt", "monaLisa (vs OtherBins)"="ML",
                 "chromVAR::differentialDeviations"="CVoriginal"), FUN=function(x){
  readRDS(paste0("../fullFrags/GATA1/runATAC_results/with_pvalues/",x,".rds"))                 
})
ints <- intersect(c("GATA1",ints$GATA1 ), unique(unlist(lapply(resl, row.names))))
resl[["optimal possible ranking"]] <- data.frame(row.names=head(c(ints, 1:100),100), rank=1:100)

d <- dplyr::bind_rows(lapply(resl, FUN=function(e){
  e$isCoFactor <- row.names(e) %in% ints
  e$propCoFact <- cumsum(e$isCoFactor)/seq_len(nrow(e))
  head(e[,c("rank","isCoFactor","propCoFact")],100)
}), .id="Method")

d$Method <- factor(d$Method, c("optimal possible ranking", "monaLisa (vs OtherBins)",
                                sort(grep("chromVAR",names(resl),value=TRUE))))
cols <- setNames(c("lightgrey", "darkgoldenrod", "deeppink2", "darkblue"),
                 levels(d$Method))

p3 <- ggplot(d, aes(x=rank, y=propCoFact, ymax=propCoFact, fill=Method, colour=Method)) + 
  geom_ribbon(ymin=0, alpha=0.4, linewidth=0.8) + theme_bw() +
  scale_fill_manual(values=cols) + scale_colour_manual(values=cols) +
  coord_cartesian(expand=0) + 
  theme(title=element_text(size=10), legend.position=c(0.62,0.76),
        panel.grid=element_blank()) +
  labs(x="Ranked motifs according to method",
       y="Proportion of top motifs among network members",
       title="Network score")


res <- readRDS("../fullFrags/results.rds")
res2 <- filterMethods(res)
p4 <- ggplot(res, aes(rank, relAUC, shape=dataset, color=method)) + 
  geom_point(show.legend = FALSE) + scale_x_log10() + theme_bw() +
  labs(x="Rank of true TF", y="Network score (relative AUC)")

```

```{r, fig.width=8, fig.height=8}
plot_grid(plot_grid(p1,p2,nrow=1,labels="AUTO",scale=0.95),
          plot_grid(p3,p4,nrow=1,labels=LETTERS[3:4],scale=0.95),
          nrow=2, rel_heights = c(4.5,5))
```


**`r getFigNb()`** 
**A: Reproducibility of chromVAR-based differential accessibility ananylsis.**
Each color represents a dataset. The line and shaded area respectively represent 
the mean and standard error of the pairwise correlations across 7 independent 
runs of the chromVAR(z-score)>limma pipeline for each dataset and setting, of the ranks 
of the (union of) top 20 motifs.
**B: Illustration of the motif-specific position weights for the insertionModel method.**
**C: Illustration of the computation of the network score.** 
The motifs are ranked for each method, and for each k=1:100 the proportion of 
top k motifs that are part of the known network of the true TF is computed. 
The area under the curve is then reported, relative to the AUC of the best 
possible motif ordering for that factor (gray area). The example plotted here is
from the GATA1 dataset. 
**D: Agreement between the two rank-based metrics.**
Each point represents the results of a method (colour) in a dataset (shape).
The network score is not entirely comparable across datasets. 
Beside this effect, and except at very low (1-3) or high (>100) ranks (i.e. 
where the ranks stop being discriminatory), there is a good agreement between
the two metrics.


\newpage


# `r getFigNb(TRUE)`

```{r}
source("../Scripts/plot.R")
source("../Scripts/runMethods.R")
res <- readRDS("../fullFrags/results.rds")
res <- renameMethods(res)

aTheme <- theme_bw() + theme(legend.position ="none", title=element_text(size=10),
                             plot.title=element_text(size=10.5))

diff2Type <- function(x){
  factor(sign(x), c("-1","0","1"), c("worsened","same","improved"))
}


CVlimma <- res[which(res$method %in% c("chromVAR(deviations)>limma","chromVAR:differentialDeviations")),]
cv1 <- reshape2::dcast(CVlimma, formula = dataset~method, value.var = "rank")
cv1$diffsign <- diff2Type(cv1$`chromVAR:differentialDeviations`-cv1$`chromVAR(deviations)>limma`)
typecols <- setNames(c("red","black","darkblue"),levels(cv1$diffsign))

# p1 <- ggplot(ranks, aes(TMM, GCsmoothQuantile, label=dataset, colour=type)) +
#   geom_abline(slope=1, colour="grey", linetype="dashed") + geom_point() + 
#   ggrepel::geom_text_repel(min.segment.length=0, size=2.5, alpha=0.6) + 
#   scale_colour_manual(values=typecols) +
#   facet_wrap(~method) + scale_y_log10() + scale_x_log10() + theme_bw() +
#   theme(legend.position = "bottom") + 
#   labs(x="Rank of true motif, using standard (TMM) normalization",
#        y="Rank of true motif, using smooth quantile normalization",
#        colour="Change in using smooth quantile normalization:") +
#   guides(colour=guide_legend(override.aes=list(size=3)))

p1 <- ggplot(cv1, aes(`chromVAR:differentialDeviations`, `chromVAR(deviations)>limma`, label=dataset, colour=diffsign)) + 
        geom_abline(slope=1, colour="grey") + geom_point() + 
        ggrepel::geom_text_repel(min.segment.length=0, size=3, alpha=0.7) + 
        scale_x_log10() + scale_y_log10() + aTheme +
        scale_colour_manual(values=typecols) +
        labs(x="Rank of true motif in chromVAR::differentialDeviations",
             y="Rank of true motif in\nchromVAR(deviations)>limma") +
        ggtitle("chromVAR>limma outperforms native\nchromVAR differential analysis")

CVlimma <- res[which(res$method %in% c("chromVAR(deviations)>limma","chromVAR(z)>Qt>limma")),]
cv1 <- reshape2::dcast(CVlimma, formula = dataset~method, value.var = "rank")
cv1$diffsign <- diff2Type(cv1$`chromVAR(deviations)>limma`-cv1$`chromVAR(z)>Qt>limma`)
p2 <- ggplot(cv1, aes(`chromVAR(deviations)>limma`, `chromVAR(z)>Qt>limma`, label=dataset, colour=diffsign)) + 
        geom_abline(slope=1, colour="grey") + geom_point() + 
        ggrepel::geom_text_repel(min.segment.length=0, size=3, alpha=0.7) + 
        scale_x_log10() + scale_y_log10() + aTheme +
        scale_colour_manual(values=typecols) +
        labs(x="Rank of true motif in chromVAR(deviations)>limma",
             y="Rank of true motif in\nchromVAR(z)>Qt>limma") +
        ggtitle("Differential analysis on normalized z-scores\nis preferable to using the adjusted deviations")
```

```{r}
dev1 <- readRDS("/mnt/bohacek/externData/2022_Caradonna_GSE200670_stress_ATAC/motifs_deviations.rds")
dev2 <- readRDS("/mnt/bohacek/externData/2022_Caradonna_GSE200670_stress_ATAC/aligned/resizedPeaks.sampleSubset.dev.SE.rds")
colnames(dev2) <- gsub("\\.bam","",colnames(dev2))
dev1 <- dev1[,colnames(dev2)]
dev2$condition <- dev1$condition
d <- dplyr::bind_rows(list(
  "Called peaks"=sechm::meltSE(dev1, row.names(dev1), assayName="z"),
  "Resized peaks"=sechm::meltSE(dev2, row.names(dev2), assayName="z")
  ), .id="type")
d$type <- factor(d$type, unique(d$type))
d2 <- sechm::meltSE(dev2, row.names(dev2), assayName="deviations")
d2$type <- "Resized peaks (deviations)"
d <- d[,c("type","sample","condition","z")]
d2 <- d2[,c("type","sample","condition","deviations")]
row.names(d) <- NULL
row.names(d2) <- NULL
d2$sample <- factor(d2$sample)
d$sample <- factor(d$sample)
saveRDS(list(d=d, d2=d2), "../misc/example_distribution_shifts.rds")

p3a <- ggplot(d, aes(z, group=sample, colour=condition)) + 
  geom_vline(xintercept=0) + 
  geom_density(show.legend = FALSE, linewidth=1) + facet_wrap(~type, nrow=2) + 
  theme_bw() + coord_cartesian(xlim=c(-20,20))
# p3b <- ggplot(d2, aes(deviations, group=sample, colour=condition)) + 
#   geom_vline(xintercept=0) + 
#   geom_density(show.legend = FALSE, linewidth=1) + facet_wrap(~type, nrow=1) + 
#   theme_bw() + coord_cartesian(xlim=c(-0.05,0.05))
# p3 <- plot_grid(p3a,p3b,nrow=1,rel_widths = c(2,1), labels=c("A","B"), scale=0.95)
```


```{r, fig.width=8, fig.height=8}
res <- readRDS("../fullFrags/results.rds")
res <- res[grep("chromVAR",res$method),]
h1 <- rankHeatmap2(res, rowann = NA, cellLabelFontsize=6, colann=NA, doDraw=FALSE,
                   row_names_gp=gpar(fontsize=9), column_names_gp=gpar(fontsize=10))

fl <- system("ls ../fullFrags/*/runATAC_results/with_pvalues/CV*", intern = TRUE)
fl <- fl[grep("CVqt|CVnorm|CVcentered|CVdev|CV\\.rds",fl)]
fl <- fl[grep("trend|ihw",fl,invert = TRUE)]
fl <- fl[grep("_1h",fl,invert=TRUE)]
ds <- basename(dirname(dirname(dirname(fl))))
d <- data.frame(Dataset=ds, method=gsub("\\.rds","",basename(fl)))
cv2 <- lapply(fl, readRDS)
datasets <- getDatasets()
d$t <- sapply(seq_along(cv2), FUN=function(i){
  x <- cv2[[i]]
  truth <- datasets[[ds[[i]]]]$truth
  x[head(which(row.names(x) %in% truth),1),"t"]
})
d$rank <- sapply(seq_along(cv2), FUN=function(i){
  x <- cv2[[i]]
  truth <- datasets[[ds[[i]]]]$truth
  x[head(which(row.names(x) %in% truth),1),"rank"]
})
direction <- ifelse(sapply(datasets, FUN=function(x) x$type)[d$Dataset]=="ligand",1,-1)
d$t <- d$t*direction
d$t[d$t<0] <- 0
d <- renameMethods(d)

d2 <- reshape2::dcast(d, Dataset~method, value.var="t")
row.names(d2) <- d2[,1]
tvals <- t(d2[,-1])
t2 <- t((t(tvals)-colMeans(tvals))/colMeans(1+tvals))
cf <- function(j, i, x, y, width, height, fill){
  grid.text(sprintf("%1.0f", tvals[i, j]), x, y,
                      gp=gpar(fontsize=6))
}
h2 <- Heatmap(t2, cell_fun = cf, cluster_rows = FALSE, cluster_columns = FALSE, row_names_side="left",
              column_title="t-values", column_order = names(h1@column_order),
              row_order = order(-rowMeans(t2)), name="Relative\ndeviation\nto dataset\naverage",
              row_names_gp=gpar(fontsize=9), column_names_gp=gpar(fontsize=10))
```


```{r, fig.width=8, fig.height=7.8}
plot_grid(plot_grid(p1, p2, nrow=1, labels=LETTERS[1:2], scale=0.9),
          plot_grid(p3a, 
                    plot_grid(grid.grabExpr(draw(h1)), grid.grabExpr(draw(h2)), nrow=2, rel_heights = c(12,11), labels=c("D","E")),
                    scale=0.95, labels=c("C",NA), nrow=1, rel_widths=c(1.5,3)),
          nrow=2, rel_heights=c(1.3,3))
```

**`r getFigNb()`: Impact of modifications to the chromVAR-limma workflow.** 
**A:** Comparison of the true motif ranks obtained by a *limma* analysis 
on the chromVAR deviations, versus using chromVAR's native 
*differentialDeviations*.
**B:** Comparison of the true motif ranks obtained by a *limma* analysis on the 
normalized chromVAR z-scores, versus on the chromVAR deviations. In **A-B**, 
datasets in blue indicate an improvement over the x-axis. Axes are 
squareroot-transformed for ease of visualization.
**C:** Example of a subset of ATAC samples from 
<a href="https://doi.org/10.1016/j.ynstr.2022.100475">Caradonna, Paul and Marrocco (2022)</a> 
(<a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE200670">GSE200670</a>),
where different samples show different shapes of z-scores distributions across motif.
When peaks are not resized to all have the same width, as recommended, the 
distributions are often globally shifted (**C**, top). Even when peaks are 
resized (**C**, bottom), differences in the width of the z-scores distributions 
which are not related to experimental groups can persist in some datasets. This 
is not specific to the z-scores, but also present in the bias-corrected 
deviations.
**D-E:** Effect on the rank of the true TF (**D**) and its *t*-value (**E**) of 
normalizing the chromVAR activity scores (i.e. z-scores or deviations) 
distributions across samples before running limma. In **E**, for datasets where 
a downregulation was expected, the *t*-values were inverted. The heatmap colors
show the method's deviation relative to the dataset average, while the numbers
indicate the actual (rounded) *t*-value.



\newpage



# `r getFigNb(TRUE)`


```{r}
dev <- readRDS("../singleCell/dev.SE.rds")
assays(dev)$scaled <- scale(assays(dev)$z)
assays(dev)$centered <- t(t(assays(dev)$z)-colMedians(assays(dev)$z))
q <- preprocessCore::normalize.quantiles(assays(dev)$z)
dimnames(q) <- dimnames(assays(dev)$z)
assays(dev)$quantile <- q

dev <- dev[,!is.na(dev$sgAssign2) & dev$sgAssignD!="D3:UNK"]
dev$TF <- gsub("sg","",dev$sgAssign3)
names(tfs) <- tfs <- setdiff(unique(dev$TF), "NT")
names(ass) <- ass <- c("z","scaled","centered","quantile")
d <- dplyr::bind_rows(lapply(ass, FUN=function(a){
  zsd <- sapply(split(seq_len(ncol(dev)),dev$sgAssign2), FUN=function(x) matrixStats::rowVars(assays(dev)[[a]][,x]))
  zm <- sapply(split(seq_len(ncol(dev)),dev$sgAssign2), FUN=function(x) rowMeans(assays(dev)[[a]][,x]))
  zm2 <- sapply(split(seq_len(ncol(zm)),gsub("sg|_[0-9]$","",colnames(zm))), FUN=function(i) rowMeans(zm[,i]))
  row.names(zm2) <- row.names(zm) <- row.names(zsd) <- row.names(dev)
  dplyr::bind_rows(lapply(tfs, FUN=function(x){
    psd <- sqrt(mean(zsd[x,grep(paste0(x,"|NT"),colnames(zsd))]))
    smd <- sapply(grep(x, colnames(zm), value=TRUE), FUN=function(i){ zm2[x,"NT"]-zm[x,i]/psd })
    data.frame(sample=names(smd), smd=as.numeric(smd))
  }), .id="TF")
}), .id="assay")
d$assay <- factor(d$assay, c("z","centered","scaled","quantile"))
dbase <- d[d$assay=="z",]
dbase <- setNames(dbase$smd, dbase$sample)
d$baseline <- dbase[d$sample]
d$difference <- d$smd-d$baseline
```


```{r, fig.width=8, fig.height=7}
d$transformation <- d$assay
levels(d$transformation)[1] <- "none"
ggplot(d, aes(transformation, smd, group=sample, colour=TF)) + 
    geom_hline(data=d[d$transformation=="z",],aes(yintercept=smd), linetype="dashed", colour="grey") +
    geom_point(size=2) + geom_line(linewidth=1) + theme_bw() + 
    labs(y="Standardized mean difference") + 
    facet_wrap(~TF, scales="free") +
    theme(legend.position="none", axis.text.x=element_text(angle=45, hjust=1, vjust=1))
```

**`r getFigNb()` : Impact of chromVAR motif z-score normalization at the single-cell level.** 
As a proxy to signal-to-noise ratio, we computed the standardized mean 
difference of the respective motif for each guide RNA (individual lines) to the 
mean of the control guide RNAs. Centering is always beneficial (Wilcoxon 
p<`r format(wilcox.test(d$difference[d$assay=="centered"], alternative="greater")$p.value, digits=1)`),
unit-variance scaling is often beneficial (Wilcoxon 
p~`r format(wilcox.test(d$difference[d$assay=="scaled"], alternative="greater")$p.value, digits=1)`) 
but sometimes detrimental, and quantile normalization also tends to be beneficial 
(Wilcoxon p~`r format(wilcox.test(d$difference[d$assay=="quantile"], alternative="greater")$p.value, digits=2)`).


\newpage

# `r getFigNb(TRUE)`


```{r, fig.width=8, fig.height=9}
res1 <- readRDS("../fullFrags/results.rds")
res2 <- readRDS("../GCnorm/results.rds")
res2$method <- paste0("GC>",res2$method)
res1$norm <- "TMM"
res2$norm <- "GCsmoothQuantile"
res <- rbind(res1[,colnames(res2)],res2)
res$method <- gsub("GCsmooth","GC",res$method)

h1 <- rankHeatmap2(res, doDraw=FALSE, column_title="Rank of true TF",
                   row_names_gp=gpar(fontsize=8.5), cellLabelFontsize=7)
h2 <- suppressWarnings(relAUCplot2(res, row_order=names(row_order(h1)), doDraw=FALSE,
                  column_title="Network score", column_order=names(column_order(h1)),
                  cellLabelFontsize=7))
h <- grid.grabExpr(draw(h1 + h2, merge=TRUE))
plot_grid(h, scale=0.95)
```

**`r getFigNb()` : Rank-based metrics of all variants, including smooth quantile normalization in GC bins.**



\newpage

# `r getFigNb(TRUE)`

```{r, fig.width=7.5, fig.height=8}
res1 <- readRDS("../fullFrags/results.rds")
res1$match <- "default"
res1 <- res1[!grepl("scale|center|deviations\\)>Qt>|fewer|smaller|spearman",res1$method),]
res2 <- readRDS("../stringentMatch/results.rds")
res2$match <- "stringent"
im <- intersect(unique(res1$method),unique(res2$method))
res1 <- res1[res1$method %in% im,colnames(res2)]
res2 <- res2[res2$method %in% im,]
co <- getOrder(res1, "transRank", columns = TRUE)
h <- rankHeatmap2(res1, rowann=NA, row_order=rev(getOrder(res1, "transRank")), column_order = co,
                   row_names_gp=gpar(fontsize=10), row_names_max_width=unit(8,"cm"), doDraw = FALSE, row_title="Default matching") %v%
rankHeatmap2(res2, rowann=NA, row_order=rev(getOrder(res2, "transRank")), column_order = co, 
                   row_names_gp=gpar(fontsize=10), row_names_max_width=unit(8,"cm"), doDraw = FALSE, row_title="Stringent matching")

res <- rbind(filterMethods(res1), filterMethods(res2))
res$method <- paste0(ifelse(res$match=="default","","stringent-"),res$method)
res$type <- res$match
p <- sensFDRplot(res, hull = FALSE)

plot_grid(grid.grabExpr(draw(h), wrap.grobs = TRUE), p,
          nrow=2, rel_heights=c(2,1.1), labels=LETTERS[1:2], scale=c(0.9,0.95))
```


**`r getFigNb()` : Impact of stringency in motif matching.** 
Rank of the true TF (**A**) and Precision-Recall (**B**) of selected methods,
when using a stringent motif matching vs the default.


\newpage

# `r getFigNb(TRUE)`

```{r, fig.width=7, fig.height=8}
res1 <- readRDS("../fullFrags/results.rds")
res1 <- res1[!grepl("scale|center|deviations\\)>Qt>|fewer|smaller|spearman|decouple|-lm|VIPER|ulm|BaGFoot",res1$method),]
res1$motifs <- "default"
res2 <- readRDS("../withArchetypes/results.rds")
res2$motifs <- "archetypes"
im <- intersect(unique(res1$method),unique(res2$method))
res1 <- res1[res1$method %in% im,colnames(res2)]
res2 <- res2[res2$method %in% im,]
co <- getOrder(res1, "transRank", columns = TRUE)
h <- rankHeatmap2(res1, rowann=NA, row_order=rev(getOrder(res1, "transRank")), cap=12, column_order = co,
                   row_names_gp=gpar(fontsize=10), row_names_max_width=unit(8,"cm"), doDraw = FALSE, row_title="All motifs") %v%
rankHeatmap2(res2, rowann=NA, row_order=rev(getOrder(res2, "transRank")), cap=12, column_order = co, 
                   row_names_gp=gpar(fontsize=10), row_names_max_width=unit(8,"cm"), doDraw = FALSE, row_title="Using archetypes")

res <- rbind(filterMethods(res1), filterMethods(res2))
res$method <- paste0(ifelse(res$motifs=="archetypes","arch-",""),res$method)
res$type <- res$motifs
p <- sensFDRplot(res, hull = FALSE)

plot_grid(grid.grabExpr(draw(h), wrap.grobs = TRUE), p, scale=0.95,
          nrow=2, rel_heights=c(2,1.1), labels=LETTERS[1:2])
```


**`r getFigNb()` : Impact of using motif archetypes.** 
Rank of the true TF (**A**) and Precision-Recall (**B**) of selected methods,
when using motifs archetypes instead of the full collection of motifs. Given the
difference in the size of the two sets of motifs, ranks were capped at 144 for 
the purpose of creating the heatmap.


\newpage

# `r getFigNb(TRUE)`

```{r, fig.width=8, fig.height=8}
res <- readRDS("../fullFrags/results.rds")
resFoN <- readRDS("../fullOnNucFreePeaks/results.rds")
resFoN <- renameMethods(resFoN)
resNF <- readRDS("../nucFree/results.rds")
resNF <- renameMethods(resNF)
resNFoF <- readRDS("../NFonFull/results.rds")
resNFoF <- renameMethods(resNFoF)
m <- merge(resFoN, resNF, by=c("dataset", "method"), suffix=c(".FoNF",".NF"))
m2 <- merge(res, resNFoF, by=c("dataset","method"), suffix=c("",".NFoF"))
m <- merge(m, m2, by=c("dataset","method"))
m <- m[m$method %in% c("chromVAR(z)>limma","monaLisa.vsOthers"),]
m$diffsign <- diff2Type(-sign(m$rank.FoNF-m$rank))
p3 <- ggplot(m, aes(rank, rank.FoNF, label=dataset, colour=diffsign)) + 
    geom_abline(slope=1, colour="grey") + geom_point() + 
    ggrepel::geom_text_repel(min.segment.length=0, size=3, alpha=0.7, show.legend=FALSE) + 
  facet_wrap(~method) + scale_x_log10() + scale_y_log10() + aTheme + 
    scale_colour_manual(values=typecols) +
    labs(x="Rank of true motif (using all fragments)",
         y="Rank of true motif\n(all frags on NF peaks)",
         colour="Change (in y relative to x):",
         title="Specifically using nucleosome-free fragments does not generally improve results")
m$diffsign <- diff2Type(-sign(m$rank.NF-m$rank))
p4 <- ggplot(m, aes(rank, rank.NF, label=dataset, colour=diffsign)) + 
    geom_abline(slope=1, colour="grey") + geom_point() + 
    ggrepel::geom_text_repel(min.segment.length=0, size=3, alpha=0.7) + 
  facet_wrap(~method) + scale_x_log10() + scale_y_log10() + aTheme +
    scale_colour_manual(values=typecols) +
    labs(x="Rank of true motif (using all fragments)", y="Rank of true motif\n(only NF frags on NF peaks)")
m$diffsign <- diff2Type(-sign(m$rank.NFoF-m$rank))
p5 <- ggplot(m, aes(rank, rank.NFoF, label=dataset, colour=diffsign)) + 
    geom_abline(slope=1, colour="grey") + geom_point() + 
    ggrepel::geom_text_repel(min.segment.length=0, size=3, alpha=0.7) + 
  facet_wrap(~method) + scale_x_log10() + scale_y_log10() + aTheme +
    scale_colour_manual(values=typecols) +
    labs(x="Rank of true motif (using all fragments)", y="Rank of true motif\n(only NF fragments)")
```


```{r, fig.width=8, fig.height=7.5}
plot_grid(
  p3, p4, p5, nrow=3, scale=c(1,1,0.95,0.95,0.95), labels=LETTERS[1:3],
  rel_heights=c(1.1,1,1)
)
```

**`r getFigNb()`: Impact of using nucelosome-free fragments.** 
**A:** Comparison of the true motif ranks obtained by the top method of each 
family of approaches, when using peaks called from nucleosome-free fragments 
(but counting all fragments overlapping them).
**B:** Comparison of the true motif ranks obtained by the top method of each 
family of approaches, when using only nucleosome-free fragments.
**C:** Comparison of the true motif ranks obtained by the top method of each 
family of approaches, when counting only nucleosome-free fragments overlapping
peaks called on all fragments. In each case, the TFs in blue indicate an 
improvement, those in red a worsening, and those in black no change.

\newpage



# `r getFigNb(TRUE)`

```{r, fig.width=5, fig.height=7.5}
bml2 <- readRDS("../misc/aggregated.bm.rds")
bml2b <- readRDS("../misc/MLM_CV.rds")
bml2b$type <- "aggregation methods"
bml2 <- rbind(bml2,bml2b)
bml2$method <- gsub("GCnorm:decoupleRlimmamlm","GCnorm>MLM>limma", bml2$method, fixed=TRUE)
bml2$method <- gsub("^ML$","monaLisa",bml2$method)
bml2$method <- gsub("CVqt","chromVAR",bml2$method)
bml2$trueQ[is.na(bml2$trueQ)] <- 1
rankBreaks=c(1,6,20,50,100,200,400)
p1 <- ggplot(bml2, aes(reorder(method,-sqrt(rank)), reorder(dataset, -sqrt(rank)), fill=sqrt(rank), label=rank)) +
  geom_tile() + facet_wrap(~type, scales="free_x") + geom_text(size=3) + theme_minimal() +
  theme(panel.grid = element_blank(),
            axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  scale_fill_viridis_c(direction=-1, breaks=sqrt(rankBreaks), 
                       trans="sqrt", labels=c("top",rankBreaks[-1])) +
  labs(x="", y="", fill="rank of \ntrue motif   ")

plot_grid(p1, sensFDRplot(bml2, hull=FALSE) + 
            coord_cartesian(ylim=c(0,0.65), xlim=c(0,1), expand=FALSE),
          nrow=2, labels="AUTO", scale=0.95)
```

**`r getFigNb()` : Simple aggregation of top methods' results does not improve inferences.**
The results of the top method from each of the two families of approach (on the left) 
are compared to those of simple aggregation methods (on the right). Beside the
established Fisher's, Stouffer's and Simes' methods, a rank-based permutation approach
was tested, establishing the probability of having a sum of ranks across methods 
lower or equal to the observed one.



\newpage

```{r, child="../simulations/sfig_sim.Rmd", eval=FALSE}
```



# `r getFigNb(TRUE)`

```{r, fig.width=4, fig.height=6}
densities <- data.frame(sample=c("NFKB.1","NFKB.2","Scrambled.1","Scrambled.1","DMSO"),
                        condition=c("NFKB","NFKB","Scrambled","Scrambled","DMSO"),
                        value=c(3674.397, 4182.298, 6229.71, 6157.883, 8588.919))
densities$condition <- factor(densities$condition, c("DMSO", "Scrambled","NFKB"))
densities$value <- densities$value/densities[which(densities$sample=="DMSO"),"value"]
dag <- aggregate(densities[,3,drop=FALSE],densities[,2,drop=FALSE],FUN=mean)
pd <- ggplot(dag, aes(condition, value)) + 
  geom_col(width=0.6, fill="deepskyblue4") + 
  stat_summary(data=densities, fun.data = "mean_se", geom = "errorbar", width=0.3, linewidth=1) + 
  theme_bw() + coord_cartesian(ylim=c(0,1)) +
  labs(y="p50/NFkB1 density ratio to GAPDH,\nas a proportion to DMSO")

plot_grid( ggdraw() + draw_image("traftac_WB.png") + coord_cartesian(expand=0),
           pd, nrow=2, labels="AUTO", rel_heights=c(2,3), scale=0.95)
```


**`r getFigNb()` : Western blot confirms the downregulation of p50/NfKB.**
Western blot (**A**) and corresponding densitometry (**B**) showing a downregulation of 
p50/NFkB in HEK cells treated with NFkB TRAFTACs versus scrambled TRAFTACs.
