---
title: "01_baseline"
author: "Emanuel Sonder"
date: "06 11 2023"
output: html_document
---

```{r setup}
library(Rsamtools)
library(BiocParallel)
library(GenomicAlignments)
library(regioneR)
library(BSgenome.Hsapiens.UCSC.hg38)
library(MASS)
library(mclust)
library(Repitools)
library(data.table)

data.table::setDTthreads(2)
source("./scripts/atacSignalSimulation.R")
source("../Scripts/runMethods.R")
source("../Scripts/compileBenchmark.R")
```

```{r, paths}
outDir <- "./data/sim_data_es"
bedBaseDir <- "./data/ATAC_beds"

# Baseline samples 
bedPaths <- paste(bedBaseDir, c("ENCFF495DQP.rds", "ENCFF130DND.rds", 
                                "ENCFF447ZRG.rds", "ENCFF966ELR.rds",
                                "ENCFF358GWK.rds", "ENCFF963YZH.rds"), sep="/")
```

```{r, copy files, eval=FALSE}
file.copy("./data/ATAC_beds/ENCFF495DQP.bam_sorted.bed", "./data/data_baseline/seq_files/CTRL1.bed")
file.copy("./data/ATAC_beds/ENCFF130DND.bam_sorted.bed", "./data/data_baseline/seq_files/CTRL2.bed")
file.copy("./data/ATAC_beds/ENCFF447ZRG.bam_sorted.bed", "./data/data_baseline/seq_files/CTRL3.bed")
file.copy("./data/ATAC_beds/ENCFF966ELR.bam_sorted.bed", "./data/data_baseline/seq_files/TRT1.bed")
file.copy("./data/ATAC_beds/ENCFF358GWK.bam_sorted.bed", "./data/data_baseline/seq_files/TRT2.bed")
file.copy("./data/ATAC_beds/ENCFF963YZH.bam_sorted.bed", "./data/data_baseline/seq_files/TRT3.bed")
```

Reformat & subset to chromosomes 1-22:
```{r, subset chrs, eval=FALSE}
bedFiles <- list.files("./data/data_baseline/seq_files", full.names=TRUE)
lapply(bedFiles[1:6], function(file){
  frags <- fread(file)
  frags <- subset(frags, V1 %in% paste0("chr", 1:22))
  frags[,start:=pmin(V2,V5)]
  frags[,end:=pmax(V3, V6)]
  colnames(frags)
  frags <- frags[,c("V1", "start", "end", "V7", "V8", "V10"), with=FALSE]
  setnames(frags, c("V1", "V7", "V8", "V10"), c("seqnames", "name", "score", "strand"))
  frags$strand <- "*"
  write.table(frags,
              file,
              quote=FALSE, col.names=FALSE,
              sep="\t",
              row.names=FALSE)
})
```

# Benchmark 

Run methods
```{r, run methods, eval=TRUE}
datasets <- list(list(truth="none",
                      species="h", 
                      type="sim",
                      seqStyle="UCSC",
                      peakFile="../peaks/merged_peaks.bed",
                      folder="./data/data_baseline/",
                      readType="bed"))
names(datasets) <- "data_baseline"

runAll(datasets, scriptsFolder="../../../Scripts", 
       methods=setdiff(getMethods(), c("ATACseqTFEA")),
       forceRerun=TRUE)
```

# Sessioninfo 

```{r}
print(sessionInfo())
```


