---
title: "03_gc_bias"
author: "Emanuel Sonder"
date: "06 11 2023"
output: html_document
---

```{r setup, include=FALSE}
library(Rsamtools)
library(BiocParallel)
library(GenomicAlignments)
library(regioneR)
library(BSgenome.Hsapiens.UCSC.hg38)
library(MASS)
library(mclust)
library(Repitools)
library(data.table)
library(ggplot2)

data.table::setDTthreads(2)
source("./scripts/atacSignalSimulation.R")
source("../Scripts/runMethods.R")
source("../Scripts/compileBenchmark.R")
```

```{bash engine.opts='-l', eval=FALSE}
# GC biases were computed with 
#cd ../fullFrags/ESR1/seqFiles
#bedtools bamtobed -i CTRL3.bam > CTRL1.bed 
# hg38Patch11.2bit from http://hgdownload.cse.ucsc.edu/gbdb/hg38/hg38.2bit
```

```{r, paths, eval=TRUE}
outDir <- "./data/sim_data_gc"
bedBaseDir <- "./data/ATAC_beds"
atacPeaksPath <- "/mnt/plger/esonder/R/tf_activity_benchmark/DTFAB/simulations/data/peaks/merged_peaks.narrowPeak"
haploPath <- "./data/haploinsufficiency.rds"
actPath <- "./data/activation.rds"
genome <- BSgenome.Hsapiens.UCSC.hg38

# Baseline samples 
bedPaths <- paste(bedBaseDir, c("ENCFF495DQP.rds", "ENCFF130DND.rds", 
                                "ENCFF447ZRG.rds", "ENCFF966ELR.rds",
                                "ENCFF358GWK.rds", "ENCFF963YZH.rds"), sep="/")

# ChIP-peaks
znf143PeaksDir <- "./data/ChIP_peaks/ENCFF500EWB_ZNF143.bed.gz"
mazPeaksDir <- "./data/ChIP_peaks/ENCFF250FJC_MAZ.bed.gz"
cebpbPeaksDir <- "./data/ChIP_peaks/ENCFF156OCY_CEBPB.bed.gz"
ctcfPeaksDir <- "./data/ChIP_peaks/ENCFF592UDD_CTCF.bed.gz"
```

```{r, gc bins, eval=TRUE}
# baseline gc content
baseline <- fread("./data/ATAC_beds/ENCFF495DQP.bam_sorted.bed")
baseline <- subset(baseline, V1=="chr1")
baseline[,start:=pmin(V2, V5)]
baseline[,end:=pmin(V3, V6)]
setnames(baseline, "V1", "chr")
frags <- makeGRangesFromDataFrame(as.data.frame(baseline))
seqlevelsStyle(frags) <- "UCSC"
frags$GC_content <- suppressWarnings(Repitools::gcContentCalc(frags, organism=genome))
fragsDt <- as.data.table(frags)

fragsDt[,gc_bins:=cut(GC_content, 
        breaks=seq(0,1, length.out=11), 
        include.lowest=TRUE)]
binDt <- fragsDt[,.(prob=.N/nrow(fragsDt)), by=gc_bins]
setorder(binDt, gc_bins)
saveRDS(fragsDt, "./data/gc_bias_files/frags_baseline_gc.rds")
saveRDS(binDt, "./data/gc_bias_files/baseline.rds")

write.table(binDt,
            "./data/gc_bias_files/baseline.txt",
            quote=FALSE, col.names=FALSE,
            sep="\t",
            row.names=FALSE)

# strong(er) gc bias content
# CTRL1 of ESR1 (CTRL1.bed)
strong <- fread("../fullFrags/ESR1/seqFiles/CTRL1.bed")
strong <- subset(strong, V1=="1")
strong[,chr:=paste0("chr", V1)]
strong$V1 <- strong$V4 <- NULL
setnames(strong, c("V2", "V3"), c("start", "end"))
frags <- makeGRangesFromDataFrame(as.data.frame(strong))
seqlevelsStyle(frags) <- "UCSC"
frags$GC_content <- suppressWarnings(Repitools::gcContentCalc(frags, organism=genome))
fragsDt <- as.data.table(frags)

fragsDt[,gc_bins:=cut(GC_content, 
        breaks=seq(0,1, length.out=11), 
        include.lowest=TRUE)]
binDt <- fragsDt[,.(prob=.N/nrow(fragsDt)), by=gc_bins]
setorder(binDt, gc_bins)
saveRDS(fragsDt, "./data/gc_bias_files/frags_strong_gc.rds")
saveRDS(binDt, "./data/gc_bias_files/strong.rds")

write.table(binDt,
            "./data/gc_bias_files/strong.txt",
            quote=FALSE, col.names=FALSE,
            sep="\t",
            row.names=FALSE)
```

```{r, plotting the gc biases}
# gc distributions files
gcStrong <- "./data/gc_bias_files/strong.txt" # comes from CTRL3.bam of ESR1 experiment
gcBaseline <- "./data/gc_bias_files/baseline.txt" # comes from 

strongDt <- fread(gcStrong, sep="\t", col.names=c("gc_bins", "prob"))
strongDt$type <- "strong"
baselineDt <- fread(gcBaseline, sep="\t", col.names=c("gc_bins", "prob"))
baselineDt$type <- "baseline"

gcBinDt <- rbind(strongDt, baselineDt)
gcBinDt[,bin_end:=as.numeric(as.factor(gc_bins))/10]

ggplot(gcBinDt, aes(x=bin_end,y=prob, fill=type, group=type))+
  geom_bar(stat = "identity", position = "identity", alpha=0.6)
  #facet_grid(rows=vars(type))
```

# Simulate 

Define simulation parameters
```{r, simulation parameters, eval=TRUE}
simParams <- as.data.table(expand.grid(es=c(0, 1, 3),
                                       tf=c("CTCF", "MAZ"),
                                       paradigm=c("activation", 
                                                  "haploinsufficiency"),
                                       gc_setting=c("group",  # group (+,+,+ vs -,-,-)
                                                    "intermediate_1", # intermediate_1 (+,-,- vs -,-,-)
                                                    "intermediate_2", # intermediate_2 (+,+,- vs -,+,-)
                                                    "all_baseline", #all_baseline (-,-,- vs -,-,-)
                                                    "all_strong"))) # all_strong (+,+,+ vs +,+,+)
# add varying of ATAC-peaks parameter in case of activation es: 0.5,1,3
# simParams[,vary_atac_peaks:=fifelse(paradigm=="activation" & es>0, TRUE, FALSE)]
simParams$vary_atac_peaks <- FALSE
#simParams[,es:=fifelse(paradigm=="activation" & es==3, 1, es)]
simParams[,peaks_path:=fifelse(tf=="ZNF143", znf143PeaksDir, "")]
simParams[,peaks_path:=fifelse(tf=="MAZ", mazPeaksDir, peaks_path)]
simParams[,peaks_path:=fifelse(tf=="CEBPB", cebpbPeaksDir, peaks_path)]
simParams[,peaks_path:=fifelse(tf=="CTCF", ctcfPeaksDir, peaks_path)]
simParams[,paradigm_path:=fifelse(paradigm=="activation", actPath, haploPath)]

# Fixed parameters
sampleNames <- c("group1_1", "group1_2", "group1_3",
                 "group2_1", "group2_2", "group2_3")
design <- c(1,1,1,-1,-1,-1)
which <- GRanges(Rle(paste0("chr", 1:22)), IRanges(start=1, width=536870912))

# here FLD & GC parameters will be the same across groups 
paramsGroup1 <- paramsGroup2 <- data.table()
paramsGroup1$name <- "group1"
paramsGroup2$name <- "group2"
```

Simulate
```{r, simulate, eval=FALSE}
# run sequentially 
lapply(1:nrow(simParams), function(i){
  
  # group (+,+,+ vs -,-,-), intermediate_1 (+,-,- vs -,-,-) and intermediate_2 (+,+,- vs -,+,-)
  if(simParams[i,]$gc_setting=="group")
  {
   gcBiases <- c(gcStrong, gcStrong, gcStrong,
                 gcBaseline, gcBaseline, gcBaseline) 
  }
  else if(simParams[i,]$gc_setting=="intermediate_1")
  {
    gcBiases <- c(gcStrong, gcBaseline, gcBaseline,
                 gcBaseline, gcBaseline, gcBaseline) 
  }
  else if(simParams[i,]$gc_setting=="intermediate_2")
  {
    gcBiases <- c(gcStrong, gcStrong, gcBaseline,
                 gcBaseline, gcStrong, gcBaseline) 
  }
  else if(simParams[i,]$gc_setting=="all_baseline")
  {
    gcBiases <- c(gcBaseline, gcBaseline, gcBaseline,
                  gcBaseline, gcBaseline, gcBaseline)
  }
  else if(simParams[i,]$gc_setting=="all_strong")
  {
     gcBiases <- c(gcStrong, gcStrong, gcStrong,
                   gcStrong, gcStrong, gcStrong)
  }
  
  lfcDist <- readRDS(simParams[i,]$paradigm_path)
  simData <- simAtacData(bamPaths=bedPaths,
                          chIPPeakDir=simParams[i,]$peaks_path,
                          atacPeakDir=atacPeaksPath,
                          sampleNames=sampleNames,
                          effectStrength=simParams[i,]$es,
                          gcBiases=gcBiases,
                          fracSub=1,
                          design=design,
                          paramsGroup1=paramsGroup1,
                          paramsGroup2=paramsGroup2,
                          lfcDist=lfcDist,
                          which=which,
                          genome=genome,
                          simGCBias=TRUE,
                          simFLD=FALSE,
                          varyAtacPeaks=simParams[i,]$vary_atac_peaks,
                          colNamesChIPPeaks=c("chr","start", "end",
                                              "name", "score", "strand",
                                             "signalValue", "pValue",
                                             "qValue", "peak"),
                         colNamesAtacPeaks=c("chr","start", "end",
                                             "name", "score", "strand",
                                             "signalValue", "pValue",
                                             "qValue", "peak"),
                         enrColChIPName="signalValue",
                         enrColAtacName="signalValue",
                         lfcCol="lfc",
                         annotationStyle="UCSC",
                         maxReadPerPeak=1e6,
                         equalLib=FALSE,
                         seed=42)
  
  folderName <- paste(simParams[i,]$tf, simParams[i,]$paradigm, 
                      simParams[i,]$es, simParams[i,]$vary_atac_peaks,
                      simParams[i,]$gc_setting, sep="_")
  folderName <- paste(folderName, "seq_files", sep="/")
  saveToBed(simData, outDir, folderName, saveMerged=FALSE)
  
  wd <- getwd()
  setwd(file.path(outDir, folderName))
  mergeBedsCommand <- "cat -u group1_1.bed group1_2.bed group1_3.bed > group1_merged.bed"
  system(mergeBedsCommand)
  mergeBedsCommand <- "cat -u group2_1.bed group2_2.bed group2_3.bed > group2_merged.bed"
  system(mergeBedsCommand)
  
  callPeaksCommand <- "nice -n 3 macs2 callpeak --outdir ../peaks -f BEDPE -n group1 -t group1_merged.bed"
  system(callPeaksCommand)
  
  callPeaksCommand <- "nice -n 3 macs2 callpeak --outdir ../peaks -f BEDPE -n group2 -t group2_merged.bed"
  system(callPeaksCommand)
  
  file.remove("group1_merged.bed")
  file.remove("group2_merged.bed")
  
  # merge peaks 
  #dir.create("../peaks")
  mergePeaksCommand <- "cat ../peaks/*narrowPeak | cut -f1-3 | sort -k1,1 -k2,2n | bedtools merge -i - > ../peaks/merged_peaks.bed"
  system(mergePeaksCommand)
  
  # Filter to non-sex chromosomes
  # peaks <- fread("../peaks/merged_peaks.narrowPeak")
  # peaks <- subset(peaks, V1 %in% paste0("chr", 1:22))
  # write.table(peaks,
  #             "../peaks/sub_merged_peaks.narrowPeak",
  #             quote=FALSE, col.names=FALSE,
  #             sep="\t",
  #             row.names=FALSE)

  setwd(wd)
  gc()
})
```

# Benchmark 

Run methods
```{r, run methods, eval=TRUE}
simParams[,folderName:=paste(tf, paradigm, es, vary_atac_peaks, gc_setting, sep="_")]
subSimParams <- simParams

datasets <- lapply(1:nrow(subSimParams), function(i){
  ds <- list(truth=subSimParams[i,]$tf,
             species="h", 
             type="sim",
             seqStyle="UCSC",
             peakFile="./peaks/merged_peaks.bed",
             folder=paste0("./data/sim_data_gc/", subSimParams[i,]$folderName),
             readType="bed")
  return(ds)
})
names(datasets) <- subSimParams$folderName

runAll(datasets, scriptsFolder="../../../../Scripts", 
       methods=getMethods(onlyTop=TRUE), forceRerun=TRUE)
```

# Sessioninfo 

```{r}
print(sessionInfo())
```