---
output:
  pdf_document: default
    latex_engine: xelatex
  html_document: default
---
```{r packages2}
suppressPackageStartupMessages({
  library(ggplot2)
  library(cowplot)
  library(ComplexHeatmap)
  library(data.table)
  library(SummarizedExperiment)
  library(patchwork)
  library(RColorBrewer)
  library(ggpointdensity)
  library(viridis)
  library(ggrepel)
  library(patchwork)
  library(limma)
  library(ggh4x)
  library(ggpubr)
  library(ggimage)
  library(stringr)
  library(BSgenome.Hsapiens.UCSC.hg38)
  library(png)
})
BSgenome <- BSgenome.Hsapiens.UCSC.hg38

source("../Scripts/plot.R")
source("../Scripts/runMethods.R")
source("../Scripts/compileBenchmark.R")
source("../Scripts/getpmoi.R")
cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", 
                "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```


# `r getFigNb(TRUE)`

```{r, fig.width=20, fig.height=14}
plot_grid(
  ggdraw() + draw_image("Sfig_sim_overview.png")
)
```


**`r getFigNb()`: Overview of the semi-simulations. **
**A:** 6 baseline ATAC-seq lymphoblastoid cell lines (LCLs) are divided into two groups. Differences will be introduced through downsampling of fragments in one of the groups, and eventual per-sample GC and fragment length biases are introduced. **B:** The same as show in main Fig 4A. The extent of downsampling is based on the enrichment over the input of the ChIP-seq peak, following one of two differential activity scenarios: an activation and a haploinsufficiency (see Fig 4B). The magnitude of the effect is further scaled up and down (perturbation strength). **C.** Settings used for the semi-simulated datasets. The resulting datasets can be divided into three groups, a first set where only the strength of the perturbation was varied by a factor (perturbation strength), a second and third one were additionally differences in respectively fragment length distribution and GC were introduced across samples.

\newpage

# `r getFigNb(TRUE)`


```{r, echo=FALSE, eval=TRUE}
# Motif proportion in ChIP peaks
peakZNF143Ranges <- fread("./data/ChIP_peaks/ENCFF500EWB_ZNF143.bed.gz")
peakZNF143Ranges <- subset(peakZNF143Ranges, V1 %in% paste0("chr", 1:22))
peakZNF143MotifRanges <- fread("./data/ChIP_peaks/motif_ZN143.bed")
peakZNF143MotifRanges <- subset(peakZNF143MotifRanges, V1 %in% paste0("chr", 1:22))

peakMAZRanges <- fread("./data/ChIP_peaks/ENCFF250FJC_MAZ.bed.gz")
peakMAZRanges <- subset(peakMAZRanges, V1 %in% paste0("chr", 1:22))
peakMAZMotifRanges <- fread("./data/ChIP_peaks/motif_MAZ.bed")
peakMAZMotifRanges <- subset(peakMAZMotifRanges, V1 %in% paste0("chr", 1:22))

peakCEBPBRanges <- fread("./data/ChIP_peaks/ENCFF156OCY_CEBPB.bed.gz")
peakCEBPBRanges <- subset(peakCEBPBRanges, V1 %in% paste0("chr", 1:22))
peakCEBPBMotifRanges <- fread("./data/ChIP_peaks/motif_CEBPB.bed")
peakCEBPBMotifRanges <- subset(peakCEBPBMotifRanges, V1 %in% paste0("chr", 1:22))

peakCTCFRanges <- fread("./data/ChIP_peaks/ENCFF592UDD_CTCF.bed.gz")
peakCTCFRanges <- subset(peakCTCFRanges, V1 %in% paste0("chr", 1:22))
peakCTCFMotifRanges <- fread("./data/ChIP_peaks/motif_CTCF.bed")
peakCTCFMotifRanges <- subset(peakCTCFMotifRanges, V1 %in% paste0("chr", 1:22))

propDt <- data.table(motif_prop=c(nrow(peakZNF143MotifRanges)/nrow(peakZNF143Ranges),
                                  nrow(peakMAZMotifRanges)/nrow(peakMAZRanges),
                                  nrow(peakCEBPBMotifRanges)/nrow(peakCEBPBRanges),
                                  nrow(peakCTCFMotifRanges)/nrow(peakCTCFRanges)), 
                    n_peaks=c(nrow(peakZNF143Ranges), nrow(peakMAZRanges),
                              nrow(peakCEBPBRanges), nrow(peakCTCFRanges)),
                    tf=c("ZNF143", "MAZ", "CEBPB", "CTCF"))

if(!file.exists("enrDt.rds")){
  countMats <- readRDS("countMats.rds")
  
  enrDt <- countMats[,.(motif_prop_chip=sum(overlaps_motif=="overlaps motif" & overlaps_chip=="overlaps ChIP")/sum(overlaps_chip=="overlaps ChIP"),
                        motif_prop_atac=sum(overlaps_motif=="overlaps motif")/.N,
                        n_chip_motif=sum(overlaps_motif=="overlaps motif" & overlaps_chip=="overlaps ChIP"),
                        n_atac_motif=sum(overlaps_motif=="overlaps motif")), 
                        by=c("tf", "paradigm", "es")]
  enrDt[,enr:=motif_prop_chip/motif_prop_atac]
  enrDt[,prop_atac_chip:=n_chip_motif/n_atac_motif]
  saveRDS(enrDt, file="enrDt.rds")
}else{
  enrDt <- readRDS("enrDt.rds")
}

enrDt <- merge(enrDt, propDt, by=c("tf"))
enrSubDt <- subset(enrDt, es==1 & paradigm=="activation")

nChIP <- ggplot(enrSubDt, aes(y="N", x=tf, fill=n_peaks))+
geom_tile(alpha=0.85)+
geom_text(aes(label=round(n_peaks, 2)), color="white",fontface="bold")+
scale_fill_viridis(option="D", end=0.95)+ # easier to see the white labels
guides(fill = guide_legend(title = "Number of ChIP-peaks",
                           label.position = "bottom"))+
scale_y_discrete(position="right")+
labs(tag="A")+
theme_bw()+
theme(axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      legend.position="bottom")

# rename variables
propDt <- enrSubDt[,c("tf", "paradigm", "es", 
                   "motif_prop_chip", "motif_prop_atac", "motif_prop")]
propDt <- melt(propDt, id.vars=c("tf", "paradigm", "es"), 
               value.name="proportion")
propDt[,variable:=as.character(variable)]
propDt[, variable:=fifelse(variable=="motif_prop", 
                           "ChIP-peaks", variable)]
propDt[, variable:=fifelse(variable=="motif_prop_chip", 
                           "ATAC-peaks\n overlapingChIP-peaks", variable)]
propDt[, variable:=fifelse(variable=="motif_prop_atac", 
                           "All ATAC-peaks", variable)]
propDt$variable <- factor(propDt$variable, 
                          levels=c("All ATAC-peaks",
                                   "ATAC-peaks\n overlapingChIP-peaks",
                                   "ChIP-peaks"),
                          ordered=TRUE)


propFullPlot <- ggplot(propDt, aes(y=variable, x=tf, fill=proportion))+
geom_tile(alpha=0.85)+
geom_text(aes(label=round(proportion, 2)), color="white",fontface="bold")+
scale_fill_viridis(option="D", end=0.95)+
guides(fill = guide_legend(title="Proportion with a motif",
                           label.position = "bottom"))+
scale_y_discrete(position="right")+
theme_bw()+
theme(axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      legend.position="bottom")

enrPlot <- ggplot(enrSubDt, aes(y="Enrichment", x=tf, fill=enr))+
geom_tile(alpha=0.85)+
geom_text(aes(label=round(enr, 2)), color="white", fontface="bold")+
scale_fill_viridis(option="D", end=0.95)+
guides(fill = guide_legend(title = "Enrichment of motifs\nin ATAC-peaks overlapping ChIP-peaks",
                           label.position = "bottom"))+
scale_y_discrete(position="right")+
theme_bw()+
theme(axis.title.x = element_blank(),
      axis.title.y = element_blank(),
      legend.position="bottom")

motifPlot <- nChIP / propFullPlot / enrPlot + plot_layout(heights = c(1, 4, 1))
```

```{r, eval=TRUE}
# Rank plots Enrichment - simulated data

freqAll <- readRDS("freqAll.rds")  

# Calculate enrichment
freqAll[,prop_chIP:=n_chIP/tot_chIP]
freqAll[,prop_ATAC:=n_atac_motif_total/tot_atac]
freqAll[,enr:=prop_chIP/prop_ATAC]
freqAll[,prop:=n_chIP_motif/n_atac_motif_total]

freqAll[,prop_co_peak:=n_co_peak/n_atac_motif_total]
freqAll[,prop_co_peak_chip:=n_co_peak_chip/n_chIP_motif]

freqAll[,rank:=frank(enr), by=.(tf, paradigm)]
freqAll[,percentile:=rank/.N, by=.(tf, paradigm)]

# get motif GC content by summing up G & C probabilities in ppm of motif model
motifs <- getNonRedundantMotifs(format="universal", species=c("Hsapiens"))
motGC <- lapply(motifs, function(m){
  m <- m@motif
  sum(rowSums(m[2:3,]))/sum(m)})
motGC <- data.table(GC_content=unlist(motGC), motif_id=names(motGC))
freqAll <- merge(freqAll, motGC, by="motif_id", all.x=TRUE)

# get top enriched tfs
freqTrueDt <- subset(freqAll, tf==motif_id)
freqMax <- subset(freqAll, rank==1)
freqTrueDt <- merge(freqTrueDt, freqMax, 
                    by=c("paradigm", "tf", "es"), suffixes=c("", ".y"))
freqTrueDt[,enr_diff:=enr-enr.y]
freqTrueDt <- freqTrueDt[,c("enr", "rank", "enr_diff", "tf", "paradigm", 
                            "es", "percentile", "prop",
                            "prop_co_peak", "prop_co_peak_chip", "GC_content")]
freqTop <- subset(freqAll, rank>=max(freqAll$rank)-2)
freqAll <- merge(freqAll, freqTrueDt[,c("tf", "paradigm", "rank")], 
                 by.x=c("tf", "paradigm"),
                 by.y=c("tf", "paradigm"))
setnames(freqAll, c("rank.x", "rank.y"), c("rank", "rank_true"))
freqTopAll <- subset(freqAll, rank>rank_true)

# Rename ZN143
freqAll[,tf:=fifelse(tf=="ZN143", "ZNF143", tf)]
freqTrueDt[,tf:=fifelse(tf=="ZN143", "ZNF143", tf)]
freqTop[,tf:=fifelse(tf=="ZN143", "ZNF143", tf)]

coRankSim <- ggplot(subset(freqAll, paradigm=="activation"), aes(x=percentile, y=enr))+
  geom_label_repel(subset(freqTop, paradigm=="activation" & !(motif_id %in% c("CTCF", "CEBPB"))), 
             mapping=aes(x=percentile, y=enr, label=motif_id), 
             alpha=1, color="grey", nudge_x=-0.3, nudge_y=0.3)+
  geom_label_repel(subset(freqTrueDt, paradigm=="activation"), 
             mapping=aes(x=percentile, y=enr, label=tf), 
             alpha=1, color="black", nudge_x=-0.5, nudge_y=0.2,  direction="x")+
  geom_point(aes(color=prop_co_peak))+
  geom_point(subset(freqTrueDt, paradigm=="activation"), 
             mapping=aes(color=prop_co_peak), size=2)+
  scale_color_gradientn(trans = "sqrt",
                        breaks=c(1.0,0.75,0.5,0.25,0.1,0),
                        colours=viridis(length(unique(freqAll$prop_co_peak))), 
                        name="Proportion of matches co-occuring with true motif")+
  facet_grid(cols=vars(tf))+
  ylab("Enrichment of motif\nin ATAC-peaks overlapping ChIP-peaks")+
  xlab("Enrichment quantile")+
  labs(tag="B")+
  theme_bw()+
  theme(legend.box.just="left",
        legend.position="bottom", 
        legend.spacing.x=unit(0.5, 'cm'),
        legend.key.width = unit(1, "cm"))

coRankGC <- ggplot(subset(freqAll, paradigm=="activation"), aes(x=percentile, y=enr))+
  geom_label_repel(subset(freqTop, paradigm=="activation" & !(motif_id %in% c("CTCF", "CEBPB"))), 
             mapping=aes(x=percentile, y=enr, label=motif_id), 
             alpha=1, color="grey", nudge_x=-0.3, nudge_y=0.3)+
  geom_label_repel(subset(freqTrueDt, paradigm=="activation"), 
             mapping=aes(x=percentile, y=enr, label=tf), 
             alpha=1, color="black", nudge_x=-0.5, nudge_y=0.2,  direction="x")+
  geom_point(aes(color=GC_content))+
  geom_point(subset(freqTrueDt, paradigm=="activation"), 
             mapping=aes(color=GC_content), size=2)+
  scale_color_viridis(option="B", name="Motif GC content")+
  facet_grid(cols=vars(tf))+
  ylab("Enrichment of motif\nin ATAC-peaks overlapping ChIP-peaks")+
  xlab("Enrichment quantile")+
  scale_size(range=c(0,1))+
  labs(tag="C")+
  theme_bw()+
  theme(legend.box.just = "left", 
        legend.margin = margin(l=0),
        legend.position="bottom", 
        legend.spacing.x=unit(0.5, 'cm'),
        legend.key.width = unit(1, "cm"))

rankSim <- coRankSim / coRankGC 

# get gc content of peaks

peakZNF143Ranges <- fread("./data/ChIP_peaks/ENCFF500EWB_ZNF143.bed.gz", 
                    select=1:3, col.names=c("chr", "start", "end"))
peakZNF143Ranges <- subset(peakZNF143Ranges, chr %in% paste0("chr", 1:22))
peakZNF143Ranges <- makeGRangesFromDataFrame(as.data.frame(peakZNF143Ranges))
peakMAZRanges <- fread("./data/ChIP_peaks/ENCFF250FJC_MAZ.bed.gz", 
                    select=1:3, col.names=c("chr", "start", "end"))
peakMAZRanges <- subset(peakMAZRanges, chr %in% paste0("chr", 1:22))
peakMAZRanges <- makeGRangesFromDataFrame(as.data.frame(peakMAZRanges))
peakCEBPBRanges <- fread("./data/ChIP_peaks/ENCFF156OCY_CEBPB.bed.gz", 
                    select=1:3, col.names=c("chr", "start", "end"))
peakCEBPBRanges <- subset(peakCEBPBRanges, chr %in% paste0("chr", 1:22))
peakCEBPBRanges <- makeGRangesFromDataFrame(as.data.frame(peakCEBPBRanges))
peakCTCFRanges <- fread("./data/ChIP_peaks/ENCFF592UDD_CTCF.bed.gz", 
                    select=1:3, col.names=c("chr", "start", "end"))
peakCTCFRanges <- subset(peakCTCFRanges, chr %in% paste0("chr", 1:22))
peakCTCFRanges <- makeGRangesFromDataFrame(as.data.frame(peakCTCFRanges))

gcPeakCEBPB <- Repitools::gcContentCalc(peakCEBPBRanges, BSgenome)
gcPeakCTCF <- Repitools::gcContentCalc(peakCTCFRanges, BSgenome)
gcPeakMAZ <- Repitools::gcContentCalc(peakMAZRanges, BSgenome)
gcPeakZNF143 <- Repitools::gcContentCalc(peakZNF143Ranges, BSgenome)
peakGC <- data.table(tf=c(rep("CEBPB", length(gcPeakCEBPB)),
                          rep("CTCF", length(gcPeakCTCF)),
                          rep("MAZ", length(gcPeakMAZ)),
                          rep("ZNF143", length(gcPeakZNF143))),
                     gc_content=c(gcPeakCEBPB, gcPeakCTCF, 
                                  gcPeakMAZ, gcPeakZNF143))

# add motif GC content annotation
freqTopAllMed <- subset(freqTopAll, paradigm=="activation")
freqTopAllMed <- freqTopAllMed[,.(GC_content=median(GC_content, na.rm=TRUE)), 
                               by=c("tf")]
freqTopAllMed[,tf:=fifelse(tf=="ZN143", "ZNF143", tf)]
freqTopAllMed$type <- "Median GC content of enriched motifs"
freqTrueDt$type <- "GC content of true motif"
gcAnnot <- rbind(freqTopAllMed, freqTrueDt[,c("tf", "GC_content", "type")])

lineTypes <- c("Median GC content of enriched motifs"="dashed",
               "GC content of true motif"="dotted")
peakGCPlot <- ggplot(peakGC, aes(x=gc_content, color=tf, fill=tf))+
geom_density(alpha=0.7)+
scale_color_manual(values=cbbPalette[2:5])+
scale_fill_manual(values=cbbPalette[2:5])+
facet_grid(cols=vars(tf))+
geom_vline(data=gcAnnot, aes(xintercept=GC_content, linetype=type))+
scale_linetype_manual(name="",values=lineTypes)+
guides(fill = guide_legend(title="Transcription Factor"))+
guides(fill="none")+
guides(color="none")+
xlab("GC content ChIP-peaks")+
labs(tag="D")+
theme_bw()+
theme(legend.position="bottom",
      legend.margin=margin(l=0, r=0, t=0, b=0))
```

```{r s2, fig.width=13, fig.height=10, eval=TRUE}
rightP <- ((coRankSim/coRankGC/peakGCPlot)+plot_layout(heights=c(3,3,2), ncol=1)) & 
  theme(plot.tag = element_text(face="bold", size=16))
plot_grid(motifPlot & theme(plot.tag = element_text(face="bold", size=16)), rightP, nrow=1, rel_widths=c(3,4))
```


**`r getFigNb()`: Overview of the simulated TFs.** Data shown correspond to the semi-simulated datasets with a pertubation strength of 1 in the activation paradigm.
**A:** Number of peaks, proportion of ChIP-seq and/or ATAC-seq peaks with a motif, and fold-enrichment of ATAC-peaks overlapping ChIP-seq peaks (i.e. bound peaks) for the motif across the four TFs. **B-C:** Enrichment of all tested motifs among the ATAC-seq peaks overlapping ChIP-seq peaks (relative to all ATAC-seq peaks), with the true motif labeled in black. The dots are colored either by co-occurence with the true motif (**B**) or by GC content, i.e. the summed probabilities of G and C across positions in the position probability matrix divided by the total sum (**C**). **D:** GC content distributions of the ChIP-seq peaks of each TF. With dotted lines the GC content of the true motif is indicated, with dashed lines the median GC content of the motifs showing stronger enrichment in the ChIP-seq peaks than the true motif.


\newpage


# `r getFigNb(TRUE)`

```{r, peak counts, eval=TRUE}
countMats <- readRDS("countMats.rds")
countMats <- subset(countMats, tf=="CEBPB" & es %in% c(0,1,3))

countPlot <- ggplot(countMats, aes(x=log2(count), color=sample)) +
  geom_density() + facet_nested(paradigm+tf~es) +
  scale_color_manual(values=c(brewer.pal(5, "Reds")[3:5], brewer.pal(5, "Blues")[3:5])) +
  labs(tag="A") + theme_bw()

countMats[,mean_group:=mean(count), 
          by=c("peak_id", "tf", "paradigm", "es", "varied_atac", "group")]
countMats[,mean_group1:=fifelse(group=="group1", mean_group, as.numeric(NA))]
countMats[,mean_group2:=fifelse(group=="group2", mean_group, as.numeric(NA))]

meanCounts <- countMats[,.(var=var(count),
                           mean=mean(count),
                           logFC=log2(max(mean_group1, na.rm=TRUE)/max(mean_group2, na.rm=TRUE)),
                           overlaps_motif=unique(overlaps_motif),
                           overlaps_chip=unique(overlaps_chip)), 
                        by=c("peak_id", "tf", "paradigm", "es", "varied_atac")]

set.seed(123)
meanCounts <- meanCounts[,.SD[sample(peak_id, 1.2e5)], by=.(paradigm, es, tf)]

# Mean count vs variance (stratified by effect strength & paradigm)
meanVarPlot <- ggplot(meanCounts, aes(x=log10(mean), y=log10(var)))+
               facet_nested(paradigm+tf~es)+
               xlab("Log10(mean count)")+
               ylab("Log10(var)")+
               ggrastr::geom_point_rast(size=0.5)+ geom_density2d() +
               labs(tag="B")+
               theme_bw()
```


```{r meanvar, fig.width=14, fig.height=8, eval=TRUE}
(countPlot / meanVarPlot ) + 
  plot_annotation(tag_levels=list(c("A", "B"))) &
  theme(plot.tag = element_text(face="bold", size=18),
        axis.title=element_text(size=16),
        legend.title=element_text(size=16),
        axis.text=element_text(size=12),
        legend.text=element_text(size=12))
```



**`r getFigNb()`: Distribution of fragment counts and variance of the semi-simulations.**
**A:** Per peak fragment count distributions of the original (dataset: "0") and simulated samples (datasets: "1", "3") with 1-fold and 3-fold perturbation strengths, for an example TF (CEBPB) across both paradigms. **B:** Mean-variance relationship of the original and simulated samples.


\newpage


# `r getFigNb(TRUE)`


```{r, eval=TRUE}
# MA plots
if(!file.exists("MAdat.subsampled.rds")){
  countMats <- readRDS("countMats.rds")
  countMats[,mean_group:=mean(count), 
            by=c("peak_id", "tf", "paradigm", "es", "varied_atac", "group")]
  countMats[,mean_group1:=fifelse(group=="group1", mean_group, as.numeric(NA))]
  countMats[,mean_group2:=fifelse(group=="group2", mean_group, as.numeric(NA))]
  
  meanCounts <- countMats[,.(var=var(count),
                             mean=mean(count),
                             logFC=log2(max(mean_group1, na.rm=TRUE)/max(mean_group2, na.rm=TRUE)),
                             overlaps_motif=unique(overlaps_motif),
                             overlaps_chip=unique(overlaps_chip)), 
                          by=c("peak_id", "tf", "paradigm", "es", "varied_atac")]
  set.seed(123)
  meanCounts <- meanCounts[,.SD[sample(peak_id, 1.2e5)], by=.(paradigm, es, tf)]
  saveRDS(meanCounts, "MAdat.subsampled.rds")
}else{
  meanCounts <- readRDS("MAdat.subsampled.rds")
}

```

```{r, eval=TRUE}
colsChIP <- c(cbbPalette[2], cbbPalette[3])
names(colsChIP) <- c("no overlap", "overlaps ChIP-peak")
meanCounts[,overlaps_chip:=fifelse(overlaps_chip!="no overlap", 
                                   "overlaps ChIP-peak",
                                    as.character(overlaps_chip))]
meanCounts$overlaps_chip <- factor(meanCounts$overlaps_chip, 
                                  levels=c("no overlap", "overlaps ChIP-peak"),
                                    ordered=TRUE)

colsMotif <- c(cbbPalette[5], cbbPalette[4])
names(colsMotif) <- c("no overlap", "overlaps motif")
meanCounts$overlaps_motif <- factor(meanCounts$overlaps_motif, 
                                    levels=c("no overlap", "overlaps motif"),
                                    ordered=TRUE)
meanPointsMotif <- meanCounts[, .(mean_access=mean(mean),
                                  mean_lfc=mean(logFC[is.finite(logFC)], na.rm=TRUE)), 
                              by=.(paradigm,es,tf,overlaps_motif)]
meanPointsChIP <- meanCounts[, .(mean_access=mean(mean),
                                  mean_lfc=mean(logFC[is.finite(logFC)], na.rm=TRUE)), 
                             by=.(paradigm,es,tf,overlaps_chip)]


ovMotif <- ggplot(meanCounts, aes(x=log10(mean), y=logFC, color=overlaps_motif))+
  facet_nested(paradigm ~ tf + es, scales="free")+
  geom_density2d(size=0.5, bins=15)+
  geom_point(data=meanPointsMotif, mapping=aes(x=log10(mean_access), y=mean_lfc, color=overlaps_motif), size=3)+
  geom_hline(yintercept=2, linetype="dashed")+
  geom_hline(yintercept=0)+
  geom_vline(xintercept=1.5, linetype="dotted")+
  geom_hline(yintercept=-2, linetype="dashed")+
  scale_color_manual(values=colsMotif, name="overlaps motif:")+
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1, linewidth=3)), title="")+
  xlab("log10(Mean Accessibility)")+
  theme_bw()+
  theme(legend.position="bottom",
         axis.text.x = element_text(angle = 90,
                                   vjust = 0.5,
                                   hjust = 0.5))
ovChIP <- ggplot(meanCounts, aes(x=log10(mean), y=logFC, color=overlaps_chip))+
  facet_nested(paradigm ~ tf + es)+
  geom_density2d(size=0.5, bins=15)+
  geom_point(data=meanPointsChIP, mapping=aes(x=log10(mean_access), y=mean_lfc, color=overlaps_chip), size=3)+
  geom_hline(yintercept=2, linetype="dashed")+
  geom_hline(yintercept=0)+
  geom_vline(xintercept=1.5, linetype="dotted")+
  geom_hline(yintercept=-2, linetype="dashed")+
  scale_color_manual(values=colsChIP, name="overlaps ChIP-peak:")+
  xlab("log10(Mean Accessibility)")+
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1, linewidth=3)), title="")+
  theme_bw()+
  theme(legend.position="bottom",
        axis.text.x = element_text(angle = 90,
                                   vjust = 0.5,
                                   hjust = 0.5))
```


```{r MAplotsOV, fig.width=14, fig.height=10, eval=TRUE}
ovChIP / ovMotif
```


**`r getFigNb()`: MA plots of the simulated effects.**
For each factor, simulated paradigm and perturbation strength, the distribution of peaks log2FC of raw fragment counts between conditions as a function of mean accessibility is plotted as 2d kernel density, separating either (**A**) ATAC-seq peaks overlapping a ChIP-seq peak or not, or (**B**) ATAC-seq peaks containing a motif or not for the TF. For visibility, the peaks were subsampled to a more comparable size across the simulated datasets (1.2$\times$10^5 peaks per dataset). Colored Points indicate the means of the respective distributions.


\newpage


# `r getFigNb(TRUE)`

```{r, eval=TRUE}
# Import method outputs
rawFiles <- list.files("./data/sim_data_es", recursive=TRUE, full.names=TRUE)
rawFiles <- rawFiles[grepl("with_pvalues", rawFiles)]
rawFiles <- rawFiles[!grepl("_0_", rawFiles)]

res <- lapply(rawFiles, readRDS)
res <- lapply(res, as.data.table, keep.rownames=TRUE)
names(res) <- rawFiles

# organize               
res <- rbindlist(res, idcol="folder", use.names=TRUE, fill=TRUE)
res[,setting:=tstrsplit(folder, split="/", keep=4)]
res[,c("tf", "paradigm", "es"):=tstrsplit(setting, keep=1:3, split="_")]
res[,method:=tstrsplit(folder, keep=7, split="/")]
res[,method:=tstrsplit(method, keep=1, split=".", fixed=TRUE)]

# get baseline peformances (es==0)
baseLineFile <- list.files("./data/data_baseline", pattern="rds$", recursive=TRUE, full.names=TRUE)
baseLineFile <- baseLineFile[grepl("with_pvalues", baseLineFile) & !grepl("old", baseLineFile)]
resBase <- lapply(baseLineFile, readRDS)
resBase <- lapply(resBase, as.data.table, keep.rownames=TRUE)
names(resBase) <- baseLineFile
resBase <- rbindlist(resBase, idcol="folder", use.names=TRUE, fill=TRUE)
resBase[,method:=tstrsplit(folder, keep=6, split="/")]
resBase[,method:=tstrsplit(method, keep=1, split=".", fixed=TRUE)]

params <- as.data.table(expand.grid(tf=c("CEBPB", "CTCF", "MAZ", "ZNF143"),
                                    paradigm=c("activation", "haploinsufficiency")))
resBase <- lapply(1:nrow(params), function(i){
  tf <- params[i,]$tf
  paradigm <- params[i,]$paradigm
  resBase$tf <- tf
  resBase$es <- 0
  resBase$paradigm <- paradigm
  resBase
})
resBase <- rbindlist(resBase)
res <- rbind(res, resBase, use.names=TRUE, fill=TRUE)
res <- renameMethods(res)

# calculate ranks
res[,rank:=order(p, -abs(logFC)), , by=c("tf", "paradigm", "es", "method")]
resSub <- subset(res, rn %in% c("CEBPB", "MAZ", "ZNF143", "ZN143", "CTCF"))
resSub <- subset(resSub, rn==tf | c(rn=="ZN143" & tf=="ZNF143"))
resSub[,is_sig:=padj<=0.05]
resSub <- subset(resSub, method!="msVIPER")
resSim <- resSub



# effect strength rank curves - positive control

rawFiles <- list.files("./data/sim_data_pos_control", recursive=TRUE, full.names=TRUE)
rawFiles <- rawFiles[grepl("with_pvalues", rawFiles)]
rawFiles <- rawFiles[!grepl("_0_", rawFiles)]
resPos <- lapply(rawFiles, readRDS)
resPos <- lapply(resPos, as.data.table, keep.rownames=TRUE)
names(resPos) <- rawFiles
                    
resPos <- rbindlist(resPos, idcol="folder", use.names=TRUE, fill=TRUE)
resPos[,setting:=tstrsplit(folder, split="/", keep=4)]
resPos[,c("tf", "paradigm", "es"):=tstrsplit(setting, keep=1:3, split="_")]
resPos[,method:=tstrsplit(folder, keep=7, split="/")]
resPos[,method:=tstrsplit(method, keep=1, split=".", fixed=TRUE)]

# get effect strength 0 files
baseLineFile <- list.files("./data/data_baseline", pattern="rds$", recursive=TRUE, full.names=TRUE)
baseLineFile <- baseLineFile[grepl("with_pvalues", baseLineFile) & !grepl("old", baseLineFile)]
resBase <- lapply(baseLineFile, readRDS)
resBase <- lapply(resBase, as.data.table, keep.rownames=TRUE)
names(resBase) <- baseLineFile
resBase <- rbindlist(resBase, idcol="folder", use.names=TRUE, fill=TRUE)
resBase[,method:=tstrsplit(folder, keep=6, split="/")]
resBase[,method:=tstrsplit(method, keep=1, split=".", fixed=TRUE)]

params <- as.data.table(expand.grid(tf=c("CEBPB", "CTCF", "MAZ", "ZNF143"),
                                    paradigm=c("activation", "haploinsufficiency")))
resBase <- lapply(1:nrow(params), function(i){
  tf <- params[i,]$tf
  paradigm <- params[i,]$paradigm
  resBase$tf <- tf
  resBase$es <- 0
  resBase$paradigm <- paradigm
  resBase
})
resBase <- rbindlist(resBase)
resPos <- rbind(resPos, resBase, use.names=TRUE, fill=TRUE)
resPos <- subset(resPos, method!="msVIPER")
resPos <- renameMethods(resPos)

# relcalculate ranks
resPos[,rank:=order(p, -abs(logFC)), by=c("tf", "paradigm", "es", "method")]
resPosSub <- subset(resPos, rn %in% c("CEBPB", "MAZ", "ZNF143", "ZN143", "CTCF"))
resPosSub <- subset(resPosSub, rn==tf | c(rn=="ZN143" & tf=="ZNF143"))
resPosSub[,is_sig:=padj<=0.05]
resPosSub <- subset(resPosSub, method!="msVIPER")
resPosSub$type <- "Only motif containing ChIP-peaks"
resSub$type <- "All ChIP-peaks"

resAll <- rbind(resPosSub, resSub, fill=TRUE, check.names=TRUE)
resAll <- subset(resAll, method %in% c("chromVAR:differentialDeviations",
                                       "chromVAR(z)>limma",
                                       "chromVAR(z)>scale>limma",
                                       "GC>fastMLM>limma",
                                       "monaLisa.StabSel",
                                       "monaLisa.vsOthers"))

# sizes are actually the same test it!!!  
act <- ggplot(subset(resAll, paradigm=="activation"), aes(x=es, y=rank, group=method)) +
geom_line(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="activation"), color="grey")+
geom_point(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="activation"), color="grey")+
geom_label_repel(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="activation" & !is_sig), 
                 color="grey", aes(label=rank, size=is_sig), direction = "y", nudge_y=5, seed=43)+
geom_label_repel(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="activation" & is_sig), fontface=2, 
                 color="grey", aes(label=rank, size=is_sig), direction = "y", nudge_y=5, seed=43)+
geom_line(data=subset(resAll, type=="Only motif containing ChIP-peaks" & paradigm=="activation"))+
geom_label(subset(resAll, paradigm=="activation" & type=="Only motif containing ChIP-peaks"), 
           mapping=aes(label=rank, x=es, y=rank, color=padj, size=is_sig), label.size=1)+
geom_text(subset(resAll, is_sig & paradigm=="activation" & type=="Only motif containing ChIP-peaks"), 
          mapping=aes(label=rank, x=es, y=rank), color="black", label.size=NA, size=4.2, fontface=2)+ # size=5
scale_size_manual(values=c("TRUE"=4.2, "FALSE"=4.2), name="significant")+
guides(size=guide_legend(override.aes = list(fontface=c(1, 2))))+
facet_grid(rows=vars(tf), cols=vars(method), scales="free")+
scale_color_viridis(option="D", direction=-1, name="adjusted p-value")+
xlab("Pertubation Strength")+
ylab("Rank of true TF")+
ggtitle("Activation")+
scale_y_sqrt()+
theme_bw()+
theme(legend.position="bottom",
      legend.title=element_text(size=12),
      legend.text=element_text(size=10))

hap <- ggplot(subset(resAll, paradigm=="haploinsufficiency"), aes(x=es, y=rank, group=method)) +
geom_line(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="haploinsufficiency"), color="grey")+
geom_point(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="haploinsufficiency"), color="grey")+
geom_label_repel(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="haploinsufficiency" & !is_sig), 
                 color="grey", aes(label=rank, size=is_sig), direction = "y", nudge_y=5, seed=43)+
geom_label_repel(data=subset(resAll, type=="All ChIP-peaks" & paradigm=="haploinsufficiency" & is_sig), fontface=2, 
                 color="grey", aes(label=rank, size=is_sig), direction = "y", nudge_y=5, seed=43)+
geom_line(data=subset(resAll, type=="Only motif containing ChIP-peaks" & paradigm=="haploinsufficiency"))+
geom_label(subset(resAll, paradigm=="haploinsufficiency" & type=="Only motif containing ChIP-peaks"), 
           mapping=aes(label=rank, x=es, y=rank, color=padj, size=is_sig), label.size=1)+
geom_text(subset(resAll, is_sig & paradigm=="haploinsufficiency" & type=="Only motif containing ChIP-peaks"), 
          mapping=aes(label=rank, x=es, y=rank), color="black", label.size=NA, size=4.2, fontface=2)+ # size=5
scale_size_manual(values=c("TRUE"=4.2, "FALSE"=4.2), name="significant")+
guides(size=guide_legend(override.aes = list(fontface=c(1, 2))))+
facet_grid(rows=vars(tf), cols=vars(method), scales="free")+
scale_color_viridis(option="D", direction=-1, name="adjusted p-value")+
xlab("Pertubation Strength")+
ylab("Rank of true TF")+
ggtitle("Haploinsufficiency")+
scale_y_sqrt()+
theme_bw()+
theme(legend.position="bottom",
      legend.title=element_text(size=12),
      legend.text=element_text(size=10))
```


```{r posControlRankLines, fig.width=16, fig.height=14, eval=TRUE}
(act / hap) +plot_layout(guides = "collect") &
  theme(plot.tag = element_text(face="bold", size=18), legend.position="bottom")
```

**`r getFigNb()`: Positive control of the semi-simulations.**
Shown are ranks and significance assigned by each method to the true motif across simulation parameters (TF, paradigm and perturbation strength). In light gray are the normal simulations (as shown in main Fig 4), while dark line and colored boxes correspond to the positive controls where a difference between the groups was only introduced in the peaks overlapping a motif of the true TF. 



```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(circlize)

getOrder <- function(res, values=c("transRank","relAUC","archRelAUC"),
                     columns=FALSE){
  res <- subset(res, effect_strength>0)
  x <- -sqrt(res$rank)
  res$transRank <- 2*sqrt(exp(x)/(1+exp(x)))
  
  y <- lapply(values, FUN=function(var){
    x <- data.table::dcast(res, method~dataset, value.var=var)

    # impute
    x <- data.table::melt(x, variable.name="dataset")
    x[,c("paradigm", "es"):=tstrsplit(dataset, split="_", keep=2:3)]
    x[,med_es:=median(value, na.rm=TRUE), by=.(method,es,paradigm)]
    x[,mean_es:=mean(value, na.rm=TRUE), by=.(method,es,paradigm)]
    x[,value:=fifelse(is.na(value),pmin(med_es, mean_es), value), by=seq_len(nrow(x))]
    x$es <- x$paradigm <- x$med_es <- x$mean_es <- NULL
    x <- data.table::dcast(x, method~dataset, value.var="value")
    
    rn <- x$method
    x$method <- NULL
    x <- as.matrix(x)
    rownames(x) <- rn
    x
  })
  if(columns){
    y <- do.call(rbind, y)
    return(colnames(y)[order(colMeans(y))])
  }
  y <- do.call(cbind, y)
  row.names(y)[order(rowMeans(y))]
}

rankHeatmap2 <- function(res, rankBreaks=c(1,10,30,75,150,300,600), ..., 
                         column_title="Datasets", doDraw=TRUE, 
                         rowann=NULL,
                         rowOrder=NULL,
                         datasetInfo=data.frame(type=sapply(getDatasets(), FUN=function(x) x$type)),
                         cellLabelFontsize=8,
                         transform=TRUE,
                         annotateAUCs=FALSE,
                         paradigm=c("activation", "haploinsufficiency"),
                         showLegend=TRUE,
                         value.var="rank"){
  setorder(res, tf, effect_strength)
  
  rankMat <- data.table::dcast(res, paradigm+method~effect_strength+tf, value.var=value.var)
  methods <- rankMat$method
  paradigm <- rankMat$paradigm
  rankMat$method <- NULL
  rankMat$paradigm <- NULL
  
  rankMat <- as.matrix(rankMat)
  if(transform) rankMat <- sqrt(rankMat)
  rownames(rankMat) <- paste(methods, paradigm, sep="_")
  
  if(annotateAUCs){
    aucNsMat <- data.table::dcast(subset(res, effect_strength>0), 
                                  paradigm+method~effect_strength+tf, 
                                  value.var="relAUC")
    aucNsMat$method <- NULL
    aucNsMat$paradigm <- NULL
    aucAsMat <- data.table::dcast(subset(res, effect_strength>0), 
                                  paradigm+method~effect_strength+tf, 
                                  value.var="archRelAUC")
    aucAsMat$method <- NULL
    aucAsMat$paradigm <- NULL
    
    rownames(aucNsMat) <- rownames(rankMat)
    rownames(aucAsMat) <- rownames(rankMat)
  }
  
  ancols <- list(type=c(deletion="darkorange3", dTag="black", ligand="darkslateblue"))
  
  colAnnot <- data.table(dataset=colnames(rankMat))
  colAnnot[,c("Pertubation strength", "TF"):=tstrsplit(dataset, split="_", keep=1:2)]
  colEs <- brewer.pal(5, "Blues")
  names(colEs) <- c(0, 0.25, 0.5, 1, 3)
  tfCols <- c("darkorange3", "black", "darkslateblue", "darkolivegreen")
  names(tfCols) <- c("CEBPB", "CTCF", "MAZ", "ZNF143")
  
  rankMat <- rankMat[,colnames(rankMat)[order(colAnnot$TF)]]
  if(annotateAUCs){
    aucNsMat[,colnames(rankMat)[order(colAnnot$TF)]]
    aucAsMat[,colnames(rankMat)[order(colAnnot$TF)]]
  }
  
  setorder(colAnnot, TF)
  
  #setorder(colAnnot, tf, effect_strength)
  colan <- HeatmapAnnotation(df=as.data.frame(colAnnot[,c("TF", "Pertubation strength"), with=FALSE]),
                             col=list("TF"=tfCols,
                                      "Pertubation strength"=colEs), 
                             show_annotation_name=FALSE)

  bdist <- log10(rankBreaks[-1]-rankBreaks[-length(rankBreaks)])
  
  rann <- getMethodAnno(methods)
  if(annotateAUCs){
    meanNsAuc <- rowMeans(aucNsMat, na.rm=TRUE)
    meanAsAuc <- rowMeans(aucAsMat, na.rm=TRUE)
    
    paradigmCols <- c("#1B9E77", "#D95F02")
    names(paradigmCols) <- c("activation", "haploinsufficiency")
    rann$col$paradigm <- paradigmCols
    rann$col$`network score` <- colorRamp2(seq(min(meanNsAuc, na.rm=TRUE), 
                                           max(meanNsAuc, na.rm=TRUE), 
                                           length.out=100), magma(100))
    rann$col$`archetype score` <- colorRamp2(seq(min(meanAsAuc, na.rm=TRUE), 
                                           max(meanAsAuc, na.rm=TRUE), 
                                           length.out=100), inferno(100))
    rowAnnot <- data.table(LFCbased=rann$df$LFCbased,
                         #family=rann$df$family,
                         paradigm=paradigm,
                         `network score`=rowMeans(aucNsMat, na.rm=TRUE),
                         `archetype score`=rowMeans(aucAsMat, na.rm=TRUE))
    }
  else{
    rowAnnot <- data.table(LFCbased=rann$df$LFCbased,
                         #family=rann$df$family,
                         paradigm=paradigm)
  }
  rownames(rowAnnot) <- rownames(rankMat)
  rowann <- rowAnnotation(df=rowAnnot, 
                          col=rann$col,
                          show_legend=colnames(rann$df)=="family")
  

  if(!is.null(rowOrder))
  {
    
    if(length(paradigm)==2){
    rowOrder <- c(paste(rev(rowOrder), "activation", sep="_"),
                  paste(rev(rowOrder), "haploinsufficiency", sep="_"))}
    else{
      rowOrder <- paste(rev(rowOrder), paradigm, sep="_")
    }
  }
  else{
    rowOrder <- rownames(rankMat)[order(rowMeans(rankMat, na.rm=TRUE))]
  }

  h <- ComplexHeatmap::Heatmap(rankMat, cluster_rows=FALSE, cluster_columns=FALSE,
          cell_fun=function(j, i, x, y, width, height, fill) {
            grid.text(sprintf("%1.0f", rankMat[i, j]^2), x, y,
                      gp = gpar(fontsize=ifelse(is.na(rankMat[i, j]),6,
                                                cellLabelFontsize)))
          },
     col=viridisLite::viridis(100, direction=-1), name="Rank of\ntrue TF",
     column_title=column_title, 
     column_labels=colAnnot$`Pertubation strength`,
     row_labels=methods,
     row_order=rowOrder, 
     show_heatmap_legend=showLegend,
     column_split=colAnnot$TF,
     row_split=rowAnnot$paradigm,
     column_order=colnames(rankMat)[order(colAnnot$TF, colAnnot$`Pertubation strength`)], 
     left_annotation=rowann, row_names_side="left", top_annotation=colan,
     heatmap_legend_param=list(at=rev(sqrt(rankBreaks)),
                               labels=rev(c("top",rankBreaks[-1])),
                               break_dist=rev(bdist), legend_height=unit(3.5, "cm"),
                               labels_gp=gpar(fontsize=9)), ...)
  
  if(!doDraw) return(h)
  draw(h, merge=TRUE)
}

# summarized AUCs
res <- readRDS("res_all_effect_strength.rds")
res <- renameMethods(as.data.frame(res))
res <- as.data.table(res)
res[,method:=fifelse(grepl("diffTF", method), "diffTF(analytic)", method)]
res <- subset(res, method!="decoupleR(udt)")
ro <- getOrder(copy(res))
h1 <- rankHeatmap2(copy(res), 
                   doDraw=FALSE, 
                   column_title="Rank of true TF", 
                   rowOrder=ro,
                   row_names_gp=gpar(fontsize=10), 
                   row_title_gp=gpar(fontface="bold"),
                   annotateAUCs=TRUE)

pdf("heatmap_full.pdf", width=10, height=14)
draw(h1, heatmap_legend_side="right", 
         annotation_legend_side="bottom")
invisible(capture.output(dev.off()))
```


# `r getFigNb(TRUE)`


```{r, out.height="\\textheight"}
knitr::include_graphics("../simulations/heatmap_full.pdf")
```

**`r getFigNb()`: Heatmap of performance of methods on the semi-simulated datasets.**
As in main Fig 1, the heatmap shows the ranks the methods achieved across different simulated pertubation strengths and fold change distributions (activation and haploinsufficiency). The annotation bar to the left shows the mean network and archetype scores on pertubation strengths greater than zero.
Ranking of the methods was performed as in Fig 1 and described in the methods section using all the datasets with a pertubation strength greater than zero. Where needed imputation of missing values for the ranking was performed as previously described using only datasets with the same pertubation strength and fold change distribution. ATACseqTFEA was excluded for runtime reasons and diffTF could not be run on two of the datasets.

\newpage


# `r getFigNb(TRUE)`

```{r, eval=TRUE}
# FLD
if(!file.exists("sim_frag_data_fld.rds")){
  simfrags <- readRDS("./data/sim_data_fld/MAZ_activation_0_FALSE_group/seq_files/simfrags.rds")
  simFrags <- simfrags$sim
  simFrags[,fld:=fifelse(group=="group1", "NF-enriched", "Mono-enriched")]
  simFrags$fld <- factor(simFrags$fld, 
                         levels=c("NF-enriched", "Mono-enriched"), ordered=TRUE)
  fragDat <- simFrags[which(simFrags$sample %in% c("group1_1","group2_1")), c("fld","width","gc_content")]
  saveRDS(fragDat, "sim_frag_data_fld.rds")
}else{
  fragDat <- readRDS("sim_frag_data_fld.rds")
}

fldDist <- ggplot(fragDat, aes(x=width, color=fld, fill=fld))+
geom_density(alpha=0.7, lwd=1.5)+
scale_color_manual(values=c("#b8db52fd", "#4d9d89fd"))+
scale_fill_manual(values=c("#b8db52fd", "#4d9d89fd"))+
xlab("Fragment Length")+
guides(color=guide_legend(title = "Fragment length distributions"),
       fill="none")+
xlim(c(0,750))+
theme_bw()+
theme(legend.title = element_blank(),
      legend.position = "bottom",
      legend.margin=margin(0,0,0,0),
      legend.box.margin=margin(0,0,0,0))
```

```{r}
rawFiles <- list.files("./data/sim_data_fld", recursive=TRUE, full.names=TRUE)
rawFiles <- rawFiles[grepl("with_pvalues", rawFiles)]
res <- lapply(rawFiles, readRDS)
res <- lapply(res, as.data.table, keep.rownames=TRUE)
names(res) <- rawFiles
                    
res <- rbindlist(res, idcol="folder", use.names=TRUE, fill=TRUE)
res[,setting:=tstrsplit(folder, split="/", keep=4)]
res[,c("tf", "paradigm", "es", "fld_1", "fld_2"):=tstrsplit(setting, keep=c(1:3,5,6), split="_")]
res[,fld:=paste(fld_1, fld_2, sep="_")]
res[,method:=tstrsplit(folder, keep=7, split="/")]
res[,method:=tstrsplit(method, keep=1, split=".", fixed=TRUE)]

# relcalculate ranks
res[,rank:=order(p, -abs(logFC)), by=c("tf", "paradigm", "es", "method", "fld")]
res <- subset(res, rn %in% c("CEBPB", "MAZ", "ZNF143", "ZN143", "CTCF"))
res <- subset(res, rn==tf | c(rn=="ZN143" & tf=="ZNF143"))
res[,is_sig:=padj<=0.05]
res <- renameMethods(res)

# rename FLD types
res[,fld:=fifelse(fld=="group_NA", "grouped", fld)]
res[,fld:=fifelse(fld=="mono_enriched", "All Mononucleosome enriched", fld)]
res[,fld:=fifelse(fld=="nf_enriched", "All Nucleosome-free enriched", fld)]

# performance plot
fldHm <- ggplot(res, aes(y=method, x=es, color=rank))+
geom_point(aes(size=is_sig))+
facet_nested(paradigm+tf~ fld, scales="free", switch="y")+
#geom_text(aes(label=rank), fontface="bold", color="black")+
#scale_color_viridis(option="D", direction=-1,
#                    name="Rank of true TF",
#                    end=0.9)+ # easier to see than the very bright yellow
scale_color_gradientn(trans = "sqrt",
                      breaks=c(500,200,100,50,10),
                     colours=magma(length(unique(res$rank)),
                                     direction=-1, end=0.95), # easier to see than the very bright yellow
                     #direction=-1,
                     name="Rank of true TF")+
                     #end=0.9)+
scale_size_manual(values=c(4, 8),
                  name="significant")+
scale_y_discrete(position="right")+
theme_bw()+
theme(legend.position="bottom",
      axis.title.x=element_blank(),
      axis.title.y=element_blank(),
      legend.key.width = unit(1, "cm"),
      legend.justification="center")

d <- data.frame(x=0, y=0, image="variations_fld.png")
ovFig1 <- ggplot(d, aes(x, y)) + 
  geom_image(aes(image=image), size=0.5,by="width")+ #, asp=1.535
  scale_size_identity()+
  theme_void()

layout <- "
AAA#CCCC
BBBBCCCC"

img <- my_image <- readPNG("variations_fld.png", native = TRUE)
ovFig1 <- ggplot() + 
  scale_x_continuous(expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  theme_void() + annotation_custom(grid::rasterGrob(img), 
                    xmin = -Inf, ymin = -Inf, xmax = Inf, ymax = Inf)+
  theme(panel.background = element_rect(fill = "transparent", colour = NA),
    plot.background = element_rect(fill = "transparent", colour = NA),
    panel.grid = element_blank(),
    panel.border = element_blank(),
    plot.margin = unit(c(-1.25, 1,-1.25,-1.25), "lines"),
    panel.margin = unit(c(0,0,0,0), "lines"),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_blank(),
    axis.line = element_blank(),
    legend.position = "none",
    axis.ticks.length = unit(0, "null"),
    axis.ticks.margin = unit(0, "null"),
    legend.margin = unit(0, "null"))
```


```{r, fig.width=12, fig.height=12}
fldPlot <- (ovFig1 + fldDist + fldHm) + 
  plot_annotation(tag_levels=list(c("A", "B", "C")))+
  plot_layout(widths = c(1.5, 1.5), heights=c(1,3), design="
              12
              33") & # design=layout
  theme(plot.tag=element_text(face="bold", size=18))
fldPlot
```

**`r getFigNb()`: Robustness to simulated variations in fragment length.**
In addition to the simulated TF effect, two different fragment length distributions (**B**) were simulated by differential sampling of the fragments according to their length for the different samples as in scheme **A**. The distribution in light green corresponds to sampling mononucleosome containing fragments with higher probability, whereas for the distribution in dark green nucleosome-free fragments were sampled preferentially. The results of the methods are shown in **C**. Colors of the points correspond to the rank of the true TF obtained, point size if the corresponding adjusted p-value was found to be significant (adjusted p-value <= 0.05). 

\newpage


# `r getFigNb(TRUE)`


```{r}
if(!file.exists("sim_frag_data_gc.rds")){
  simfrags <- readRDS("./data/sim_data_gc/MAZ_activation_0_FALSE_group/seq_files/simfrags.rds")
  simFrags <- simfrags$sim
  simFrags[,gc:=fifelse(group=="group1", "alternative", "baseline")]
  simFrags$fld <- factor(simFrags$gc, 
                         levels=c("baseline", "alternative"), ordered=TRUE)
  fragDat <- simFrags[which(simFrags$sample %in% c("group1_1","group2_1")), c("gc","width","gc_content")]
  saveRDS(fragDat, "sim_frag_data_gc.rds")
}else{
  fragDat <- readRDS("sim_frag_data_gc.rds")
}

gcDist <- ggplot(fragDat, aes(x=gc_content, color=gc, fill=gc))+
geom_density(alpha=0.7, lwd=1.5)+
scale_color_manual(values=c("#952758ff", "#fa8d24f9"))+
scale_fill_manual(values=c("#952758ff", "#fa8d24f9"))+
xlab("GC content")+
guides(color=guide_legend(title="GC design"),
       fill="none")+
theme_bw()+
theme(legend.title = element_blank(),
      legend.position = "bottom")
```


```{r}
rawFiles <- list.files("./data/sim_data_gc", recursive=TRUE, full.names=TRUE)
rawFiles <- rawFiles[grepl("with_pvalues", rawFiles)]
res <- lapply(rawFiles, readRDS)
res <- lapply(res, as.data.table, keep.rownames=TRUE)
names(res) <- rawFiles
                    
res <- rbindlist(res, idcol="folder", use.names=TRUE, fill=TRUE)
res[,setting:=tstrsplit(folder, split="/", keep=4)]
res[,c("tf", "paradigm", "es", "gc_1", "gc_2"):=tstrsplit(setting, keep=c(1:3,5,6), split="_")]
res[,gc:=paste(gc_1, gc_2, sep="_")]

res[,method:=tstrsplit(folder, keep=7, split="/")]
res[,method:=tstrsplit(method, keep=1, split=".", fixed=TRUE)]

# rename GC types
res[,gc:=fifelse(gc=="all_baseline", "all baseline", gc)]
res[,gc:=fifelse(gc=="all_strong", "all alternative", gc)]
res[,gc:=fifelse(gc=="group_NA", "grouped", gc)]
res[,gc:=fifelse(gc=="intermediate_1", "intermediate", gc)]
res$gc <- factor(res$gc, levels=c("all baseline", "all alternative", 
                                  "intermediate", "grouped"))

# relcalculate ranks
res[,rank:=order(p, -abs(logFC)), by=c("tf", "paradigm", "es", "method", "gc")]
res <- subset(res, rn %in% c("CEBPB", "MAZ", "ZNF143", "ZN143", "CTCF"))
res <- subset(res, rn==tf | c(rn=="ZN143" & tf=="ZNF143"))
res[,is_sig:=padj<=0.05]
res <- renameMethods(res)

gcHm <- ggplot(res, aes(y=method, x=es, color=rank))+
geom_point(aes(size=is_sig))+
facet_nested(paradigm+tf~ gc, scales="free", switch="y")+
scale_color_gradientn(trans = "sqrt",
                      breaks=c(500,200,100,50,10),
                      colours=viridis(length(unique(res$rank)), 
                                    direction=-1, end=0.95), # easier to see than the very bright yellow
                     #direction=-1,
                     name="Rank of true TF")+
#geom_text(aes(label=rank), fontface="bold", color="black")+
scale_size_manual(values=c(4,8), name="significant")+
scale_y_discrete(position="right")+
theme_bw()+
theme(legend.position="bottom",
      legend.key.width = unit(1, "cm"),
      axis.title.x=element_blank(),
      axis.title.y=element_blank(),
      legend.justification="center")

img <- my_image <- readPNG("variations_gc.png", native = TRUE)
ovFig1 <- ggplot() + 
  scale_x_continuous(expand=c(0,0)) +
  scale_y_continuous(expand=c(0,0)) +
  theme_void() + annotation_custom(grid::rasterGrob(img), 
                    xmin = -Inf, ymin = -Inf, xmax = Inf, ymax = Inf)+
  theme(panel.background = element_rect(fill = "transparent", colour = NA),
    plot.background = element_rect(fill = "transparent", colour = NA),
    panel.grid = element_blank(),
    panel.border = element_blank(),
    plot.margin = unit(c(-1.25, 1,-1.25,-1.25), "lines"),
    panel.margin = unit(c(0,0,0,0), "lines"),
    axis.ticks = element_blank(),
    axis.text = element_blank(),
    axis.title = element_blank(),
    axis.line = element_blank(),
    legend.position = "none",
    axis.ticks.length = unit(0, "null"),
    axis.ticks.margin = unit(0, "null"),
    legend.margin = unit(0, "null"))
```


```{r, fig.width=12, fig.height=12}
gcPlot <- (ovFig1 + gcDist + gcHm) + plot_annotation(tag_levels=list(c("A", "B", "C")))+
  plot_layout(widths = c(1.5, 1.5), heights=c(1,3), design="
              12
              33") &
  theme(plot.tag=element_text(face="bold", size=18))
gcPlot
```

**`r getFigNb()`: Robustness to simulated variations in GC content.**
In addition to the simulated TF effect, two different GC content distributions (**B**) were simulated by differential sampling of the fragments according to their GC content for the different samples as in scheme **A**. The results of the methods are shown in **C**. Colors of the points correspond to the rank of the true TF obtained, point size if the corresponding adjusted p-value was found to be significant (adjusted p-value <= 0.05). 


\newpage


# `r getFigNb(TRUE)`

```{r, top methods}
topMethods <- c("monaLisa.vsOthers", # real data
                "chromVAR(z)>Qt>limma",
                "diffTF(analytic)",
                "decoupleR(mlm)>limma",
                "MEIRLOP", # both
                "msVIPER(binary)", # simulated data
                "chromVAR(z)>limma",
                "decoupleR(consensus)>limma",
                "ulm+GC")
```

```{r, fig.width=11, fig.height=6}
res1 <- as.data.table(readRDS("../fullFrags/results.rds"))
res1$type <- "real"
res2 <- readRDS("../simulations/res_all_effect_strength.rds")
res2 <- renameMethods(as.data.frame(res2))
res2 <- as.data.table(res2)
res2[,method:=fifelse(grepl("diffTF", method), "diffTF(analytic)", method)]
res2 <- subset(res2, method %in% topMethods)

res2$type <- "sim"
res2[,c("tf", "paradigm", "es"):=tstrsplit(dataset, split="_", keep=1:3)]
res1 <- res1[which(res1$method %in% res2$method),]

res <- rbind(res1, res2, use.names=TRUE, fill=TRUE)
res$FDR[which(is.na(res$FDR))] <- 0
res[,precision:=1-FDR]

# get number of significants per method for the effect strength 0 dataset
baseRes <- list.files("./data/data_baseline", recursive=TRUE, full.names=TRUE)
baseRes <- baseRes[grepl("with_pvalues", baseRes)]
baseResNames <- unlist(tstrsplit(baseRes, "/",keep=6))
baseResNames <- unlist(tstrsplit(baseResNames, split=".", keep=1, fixed=TRUE))
baseRes <- lapply(baseRes, readRDS)
baseRes <- lapply(baseRes, function(br) sum(br$padj<=0.05, na.rm=TRUE))
baseRes <- data.table(n_sig_baseline=unlist(baseRes), method=baseResNames)
baseRes[,method:=renameMethods(method)]
res <- merge(res, baseRes, 
             by.x=c("method"), 
             by.y=c("method"), all.x=TRUE, all.y=FALSE)

# there is no true positive in the case of pertubation/effect strength 0
res[,precision:=fifelse((FDR>0 |trueQ<=0.05 | n_sig_baseline>0) & 
                         effect_strength==0 & type=="sim", 0, precision)]
res[,is_sig:=trueQ<=0.05]


# rename simulated datasets
res[,dataset:=fifelse(type=="sim" & paradigm=="activation", 
                      paste(tf, "activ.", sep="\n"), dataset)]
res[,dataset:=fifelse(type=="sim" & paradigm=="haploinsufficiency", 
                      paste(tf, "haplo.", sep="\n"), dataset)]

res[,precision_color:=fifelse(precision<0.9, TRUE, FALSE)]

# order methods
orderMethods <- res[,.(mean_precision=mean(precision)), by=method]
res$method <- factor(res$method,
                     levels=orderMethods$method[order(orderMethods$mean_precision)],
                     ordered=TRUE)

# order datasets 
orderDatasets <- res[,.(mean_precision=mean(precision)), by=dataset]
res$dataset <- factor(res$dataset,
                      levels=orderDatasets$dataset[order(orderDatasets$mean_precision)],
                      ordered=TRUE)

tlr <- ggplot(subset(res, type=="real"), 
       aes(x=dataset, fill=precision, y=method))+
      scale_fill_gradientn(trans = "sqrt",
                           breaks=c(1.0,0.75,0.5,0.25,0.1,0),
                           colours=viridis(20), 
                           name="Precision")+
      labs(tag="A")+
      geom_tile()+
      geom_text(aes(label=round(precision, 2), color=precision_color),
                fontface="bold")+
      scale_color_manual(values=c("TRUE"="white", "FALSE"="black"), 
                         guide="none")+
      scale_x_discrete(guide = guide_axis(n.dodge=2))+
      theme_bw()+
      theme(axis.text.y = element_text(size=12, color="black"),
            axis.text.x = element_text(size=12, color="black"),
            plot.tag=element_text(face="bold", size=18),
            legend.position="bottom",
            axis.title.y=element_blank(),
            axis.title.x=element_blank())

# tls 
resAg <- subset(res, type=="sim")[,.(precision=mean(precision)), 
                                  by=.(es, method)]
resAg[,precision_color:=fifelse(precision<0.9, TRUE, FALSE)]
tls <- ggplot(resAg, 
       aes(x=es, fill=precision, y=method))+
       scale_fill_gradientn(trans = "sqrt",
                           breaks=c(1.0,0.75,0.5,0.25,0.1,0),
                           colours=viridis(20), 
                           name="Precision")+
      geom_tile()+
      geom_text(aes(label=round(precision, 2), color=precision_color), 
                fontface="bold")+
      scale_color_manual(values=c("TRUE"="white", "FALSE"="black"), 
                         guide="none")+
      #scale_x_discrete(guide = guide_axis(n.dodge=2))+
      labs(tag="B")+
      theme_bw()+
      theme(axis.text.y = element_blank(),
            axis.text.x = element_text(size=12, color="black", 
                                       angle=90, hjust = 1, vjust=0.5),
            plot.tag=element_text(face="bold", size=18),
            legend.position="bottom",
            axis.title=element_blank())

(tlr + ggtitle("Real datasets")) + (tls + ggtitle("Simulations")) + 
  plot_layout(guides="collect", widths=c(4,2)) & 
  theme(legend.position = "bottom", legend.key.width = unit(2,"cm"))
```

**`r getFigNb()`: Estimated precisions per dataset.**
Heatmap of precision estimates obtained per dataset and method on the 11 real datasets (**A**), as well as averaged across each perturbation strength of the simulations (**B**).